// @m-design/m-site-navbar/lib/components/tabs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-nocheck
var react_1 = __importStar(require("react"));
var web_react_1 = require("@arco-design/web-react");
var icon_1 = require("@arco-design/web-react/icon");
var classnames_1 = __importDefault(require("classnames"));
var dropdown_1 = __importDefault(require("../dropdown"));
var locale_1 = __importDefault(require("./locale"));
var index_module_less_1 = __importDefault(require("./index.module.less"));
function NavTabs(_a) {
    var _b = _a.lang, lang = _b === void 0 ? 'zh-CN' : _b, visible = _a.visible, tabs = _a.tabs, onVisibleChange = _a.onVisibleChange, getPopupContainer = _a.getPopupContainer;
    var _c = (0, react_1.useState)(''), activeKey = _c[0], setActiveKey = _c[1];
    var timer = (0, react_1.useRef)(null);
    var t = locale_1.default[lang];
    function getActiveClassNames(field) {
        var _a;
        return (0, classnames_1.default)(index_module_less_1.default.item, ((_a = {}), (_a[index_module_less_1.default.itemHover] = field === activeKey), _a));
    }
    function onMouseEnter(activeKey) {
        setActiveKey(activeKey);
    }
    return react_1.default.createElement(web_react_1.Trigger, {
        popup: function () {
            return react_1.default.createElement(dropdown_1.default, { activeModule: activeKey, lang: lang });
        },
        trigger: 'hover',
        classNames: 'slideTop',
        style: { zIndex: 998 },
        popupVisible: visible,
        onVisibleChange: function (v) {
            onVisibleChange(v);
            if (!v) {
                setActiveKey('');
            }
        },
        getPopupContainer: getPopupContainer,
    }, react_1.default.createElement('div', { className: index_module_less_1.default.tabs }, tabs.map(function (tab) {
        return react_1.default.createElement('div', {
            key: tab.title,
            className: getActiveClassNames(tab.activeKey),
            onMouseEnter: function () {
                timer.current = setTimeout(function () {
                    onMouseEnter(tab.activeKey);
                }, 200);
            },
            onMouseLeave: function () {
                clearTimeout(timer.current);
                timer.current = null;
            },
        }, react_1.default.createElement('a', { href: tab.href, className: index_module_less_1.default.itemInner, target: '_blank', rel: 'noreferrer' }, t[tab.title], tab.dropdown && react_1.default.createElement(icon_1.IconDown, null)));
    })));
}
NavTabs.defaultProps = {
    tabs: [
        {
            title: 'design',
            href: '/docs/spec/introduce',
            activeKey: 'spec',
            dropdown: true,
        },
        {
            title: 'development',
            href: '/react/docs/start',
            activeKey: 'development',
            dropdown: true,
        },
        {
            title: 'ecosystem',
            activeKey: 'ecosystem',
            dropdown: true,
        },
    ],
};
exports.default = NavTabs;
