// @m-design/m-site-footer/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-nocheck
var react_1 = __importStar(require("react"));
var icon_1 = require("@arco-design/web-react/icon");
var classnames_1 = __importDefault(require("classnames"));
// import ResizeObserver from './ResizeObserver';
var rc_resize_observer_1 = __importDefault(require("rc-resize-observer"));
// import Logo from './ArcoLogo';
var footer_module_less_1 = __importDefault(require("./style/footer.module.less"));
var locales = {
    'zh-CN': {
        design: '设计',
        component: '组件',
        ecosystem: '生态产品',
        resource: '资源',
        feedback: '反馈',
        about: '关于 Arco',
        overview: '组件索引',
        spec: '设计规范',
        principle: '设计原则',
        start: '快速开始',
        question: '常见问题',
        migrate: '从 v1 到 v2',
        changelog: '更新日志',
        pro: '中后台最佳实践',
        componentFigma: '组件库 Figma 资源',
        iconFigma: '图标 Figma 资源',
        toolWebpack: 'Webpack Plugin',
        designLab: '设计系统配置平台',
        material: '物料平台',
    },
};
function Footer(props) {
    var style = props.style, logo = props.logo, lang = props.lang, isDoc = props.isDoc;
    var t = locales[lang];
    var footerRef = (0, react_1.useRef)(null);
    var _b = (0, react_1.useState)(''), screen = _b[0], setScreen = _b[1];
    var _c = (0, react_1.useState)(''), expandKey = _c[0], setExpandKey = _c[1];
    function updateScreen() {
        var width = footerRef.current.clientWidth;
        if (width >= 1600) {
            setScreen('xxl');
        }
        else if (width >= 1200 && width < 1600) {
            setScreen('xl');
        }
        else if (width >= 992 && width < 1200) {
            setScreen('lg');
        }
        else if (width >= 768 && width < 992) {
            setScreen('md');
        }
        else if (width < 768) {
            setScreen('mobile');
        }
    }
    function handleExpand(key) {
        if (screen === 'mobile') {
            if (expandKey === key) {
                setExpandKey('');
            }
            else {
                setExpandKey(key);
            }
        }
    }
    function getListHeightStyle(key, count) {
        return screen === 'mobile' ? { height: expandKey === key ? 48 * count : 0 } : {};
    }
    function getOpenClassName(key) {
        var _a = {};
        _a["".concat(footer_module_less_1.default.open)] = key === expandKey;
        return (0, classnames_1.default)(footer_module_less_1.default.title, _a);
    }
    return (react_1.default.createElement(rc_resize_observer_1.default, { onResize: updateScreen },
        react_1.default.createElement("footer", { ref: footerRef, className: (0, classnames_1.default)(footer_module_less_1.default.footer, footer_module_less_1.default[screen], isDoc ? footer_module_less_1.default.document : ''), style: style },
            react_1.default.createElement("div", { className: footer_module_less_1.default.footerInner },
                react_1.default.createElement("div", { className: footer_module_less_1.default.content },
                    react_1.default.createElement("div", { className: footer_module_less_1.default.logo }, logo),
                    react_1.default.createElement("div", { className: footer_module_less_1.default.lw },
                        react_1.default.createElement("div", { className: footer_module_less_1.default.li },
                            react_1.default.createElement("div", { className: getOpenClassName('design'), onClick: function () { return handleExpand('design'); } },
                                t.design,
                                react_1.default.createElement(icon_1.IconDown, null)),
                            react_1.default.createElement("div", { className: footer_module_less_1.default.list, style: getListHeightStyle('design', 2) },
                                react_1.default.createElement("a", { className: footer_module_less_1.default.item, href: "/docs/spec/introduce" }, t.spec),
                                react_1.default.createElement("a", { className: footer_module_less_1.default.item, href: "/docs/spec/values-of-arcodesign" }, "\u8BBE\u8BA1\u4EF7\u503C\u89C2"))),
                        react_1.default.createElement("div", { className: footer_module_less_1.default.li },
                            react_1.default.createElement("div", { className: getOpenClassName('component'), onClick: function () { return handleExpand('component'); } },
                                t.component,
                                react_1.default.createElement(icon_1.IconDown, null)),
                            react_1.default.createElement("div", { className: footer_module_less_1.default.list, style: getListHeightStyle('component', 5) },
                                react_1.default.createElement("a", { className: footer_module_less_1.default.item, href: "/react/docs/start" }, t.start),
                                react_1.default.createElement("a", { className: footer_module_less_1.default.item, href: "/react/docs/question" }, t.question),
                                react_1.default.createElement("a", { className: footer_module_less_1.default.item, href: "/react/docs/changelog" }, t.changelog))),
                        react_1.default.createElement("div", { className: (0, classnames_1.default)(footer_module_less_1.default.li, footer_module_less_1.default.liwd) },
                            react_1.default.createElement("div", { className: getOpenClassName('ecosystem'), onClick: function () { return handleExpand('ecosystem'); } },
                                t.ecosystem,
                                react_1.default.createElement(icon_1.IconDown, null)),
                            react_1.default.createElement("div", { className: footer_module_less_1.default.list, style: getListHeightStyle('ecosystem', 7) },
                                react_1.default.createElement("a", { className: footer_module_less_1.default.item, href: "/themes/stores" }, t.designLab),
                                react_1.default.createElement("a", { className: footer_module_less_1.default.item, href: "/material" }, t.material))),
                        react_1.default.createElement("div", { className: footer_module_less_1.default.li },
                            react_1.default.createElement("div", { className: getOpenClassName('resource'), onClick: function () { return handleExpand('resource'); } },
                                t.resource,
                                react_1.default.createElement(icon_1.IconDown, null)),
                            react_1.default.createElement("div", { className: footer_module_less_1.default.list, style: getListHeightStyle('resource', 6) },
                                react_1.default.createElement("a", { className: footer_module_less_1.default.item, href: "https://github.com/arco-design/arco-webpack-plugin" }, t.toolWebpack))))),
                react_1.default.createElement("div", { className: footer_module_less_1.default.bottom },
                    react_1.default.createElement("div", { className: footer_module_less_1.default.left },
                        react_1.default.createElement("div", { className: footer_module_less_1.default.copyright },
                            react_1.default.createElement("span", null, "UED & \u524D\u7AEF "),
                            react_1.default.createElement("span", null, "Copyright Meetsocial 2021"))))))));
}
Footer.defaultProps = {
    logo: react_1.default.createElement("span", null, "M Design"),
    lang: 'zh-CN',
};
exports.default = Footer;
