// rc-tween-one/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Ticker: true,
  Plugins: true,
  Easing: true,
  TweenOneGroup: true,
  ChildrenPlugin: true,
  PathMotionPlugin: true,
  SvgDrawPlugin: true,
  SvgMorphPlugin: true
};
Object.defineProperty(exports, "Ticker", {
  enumerable: true,
  get: function get() {
    return _tweenOne.Ticker;
  }
});
Object.defineProperty(exports, "Plugins", {
  enumerable: true,
  get: function get() {
    return _tweenOne.Plugins;
  }
});
Object.defineProperty(exports, "Easing", {
  enumerable: true,
  get: function get() {
    return _tweenOne.Easing;
  }
});
Object.defineProperty(exports, "TweenOneGroup", {
  enumerable: true,
  get: function get() {
    return _TweenOneGroup.default;
  }
});
Object.defineProperty(exports, "ChildrenPlugin", {
  enumerable: true,
  get: function get() {
    return _ChildrenPlugin.default;
  }
});
Object.defineProperty(exports, "PathMotionPlugin", {
  enumerable: true,
  get: function get() {
    return _PathMotionPlugin.default;
  }
});
Object.defineProperty(exports, "SvgDrawPlugin", {
  enumerable: true,
  get: function get() {
    return _SvgDrawPlugin.default;
  }
});
Object.defineProperty(exports, "SvgMorphPlugin", {
  enumerable: true,
  get: function get() {
    return _SvgMorphPlugin.default;
  }
});
exports.default = void 0;

var _tweenOne = require("tween-one");

var _TweenOne = _interopRequireDefault(require("./TweenOne"));

var _TweenOneGroup = _interopRequireDefault(require("./TweenOneGroup"));

var _ChildrenPlugin = _interopRequireDefault(require("./plugin/ChildrenPlugin"));

var _PathMotionPlugin = _interopRequireDefault(require("./plugin/PathMotionPlugin"));

var _SvgDrawPlugin = _interopRequireDefault(require("./plugin/SvgDrawPlugin"));

var _SvgMorphPlugin = _interopRequireDefault(require("./plugin/SvgMorphPlugin"));

var _type = require("./type");

Object.keys(_type).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _type[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _type[key];
    }
  });
});
_TweenOne.default.plugins = _tweenOne.Plugins;
_TweenOne.default.ticker = _tweenOne.Ticker;
_TweenOne.default.easing = _tweenOne.Easing;
var _default = _TweenOne.default;
exports.default = _default;
