// @m-design/m-site-navbar/lib/components/search-modal/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-nocheck
var react_1 = __importStar(require("react"));
var web_react_1 = require("@arco-design/web-react");
var icon_1 = require("@arco-design/web-react/icon");
var style_module_less_1 = __importDefault(require("./style/style.module.less"));
var history_module_less_1 = __importDefault(require("./style/history.module.less"));
var hot_module_less_1 = __importDefault(require("./style/hot.module.less"));
var style_module_less_2 = __importDefault(require("../search-input/style.module.less"));
var list_module_less_1 = __importDefault(require("./style/list.module.less"));
var icon_2 = __importDefault(require("../../icon"));
var components_1 = __importDefault(require("./components"));
var data_1 = __importDefault(require("./data"));
var list_1 = __importDefault(require("./list"));
var context_1 = require("../../context");
var case_1 = require("../../utils/case");
var locale_1 = __importDefault(require("../../locale"));
var highlightStr = function (name, query) {
    var regex = new RegExp("(".concat(query.replace(/\\/g, '\\\\'), ")"), 'gi');
    return name.replace(regex, '<span class="highlight-word">$1</span>');
};
var locales = {
    'zh-CN': {
        component: '组件',
        ecosystem: '生态',
        noData: '暂无搜索记录',
        openSearch: '打开搜索',
        exit: '退出',
        recentSearch: '最近搜索',
        searchResult: '搜索到 {0} 个结果',
        clearAll: '清除全部',
        popularComponents: '热门组件',
        popularResource: '热门资源',
        button: '按钮 Button',
        table: '表格 Table',
        form: '表单 Form',
        select: '选择器 Select',
        'figma-component': '组件库设计资源',
        'design-lab': '风格配置平台',
        material: '物料平台',
        pro: 'ArcoDesign Pro',
    },
    'en-US': {
        component: 'Component',
        ecosystem: 'Ecosystem',
        noData: 'No search history',
        openSearch: 'Open Search',
        exit: 'Exit',
        recentSearch: 'Recent Search',
        searchResult: '{0} results found',
        clearAll: 'Clear All',
        popularComponents: 'Popular components',
        popularResource: 'Popular Resource',
        button: 'Button',
        table: 'Table',
        form: 'Form',
        select: 'Select',
        'figma-component': 'Figma of components',
        'design-lab': 'Design Lab',
        material: 'Arco Material',
        pro: 'ArcoDesign Pro',
    },
};
var popularComponents = [
    {
        name: 'button',
        url: '/react/components/button',
        svg: react_1.default.createElement(react_1.default.Fragment, null, react_1.default.createElement('span', { className: hot_module_less_1.default.n }, icon_2.default.IconButtonGray), react_1.default.createElement('span', { className: hot_module_less_1.default.h }, icon_2.default.IconButtonBlue)),
    },
    {
        name: 'table',
        url: '/react/components/table',
        svg: react_1.default.createElement(react_1.default.Fragment, null, react_1.default.createElement('span', { className: hot_module_less_1.default.n }, icon_2.default.IconTableGray), react_1.default.createElement('span', { className: hot_module_less_1.default.h }, icon_2.default.IconTableBlue)),
    },
    {
        name: 'form',
        url: '/react/components/form',
        svg: react_1.default.createElement(react_1.default.Fragment, null, react_1.default.createElement('span', { className: hot_module_less_1.default.n }, icon_2.default.IconFormGray), react_1.default.createElement('span', { className: hot_module_less_1.default.h }, icon_2.default.IconFormBlue)),
    },
    {
        name: 'select',
        url: '/react/components/select',
        svg: react_1.default.createElement(react_1.default.Fragment, null, react_1.default.createElement('span', { className: hot_module_less_1.default.n }, icon_2.default.IconSelectGray), react_1.default.createElement('span', { className: hot_module_less_1.default.h }, icon_2.default.IconSelectBlue)),
    },
];
var popularResource = [
    // {
    //   name: 'figma-component',
    //   url:
    //     'https://www.figma.com/file/M66cTiLXHa4SVyZIlfY5Pb/arco-Design-System?node-id=7945%3A44563',
    //   svg: icon.IconFigma,
    //   color: '#000',
    // },
    {
        name: 'design-lab',
        url: '/themes',
        svg: icon_2.default.IconLogoDesignLabW,
        color: '#165DFF',
    },
    {
        name: 'material',
        url: '/material',
        svg: icon_2.default.IconLogoArcoMaterialW,
        color: '#0099FF',
    },
    // {
    //   name: 'pro',
    //   url: '/pro',
    //   svg: icon.IconLogoArcoProW,
    //   color: '#165DFF',
    // },
];
function getBackgroundColor(svg) {
    var blueList = [
        'IconLogoArcoMaterialW',
        'IconLogoArcoProW',
        'IconLogoDesignLabW',
        'IconBranding',
        'IconDashboard',
        'IconCli',
    ];
    if (svg && (blueList.indexOf(svg) > -1 || !icon_2.default[svg])) {
        return '#165DFF';
    }
    return undefined;
}
function SearchModal(_a) {
    var visible = _a.visible, setVisible = _a.setVisible, _b = _a.lang, lang = _b === void 0 ? 'zh-CN' : _b, onChangeSearchInput = _a.onChangeSearchInput, _c = _a.queryData, queryData = _c === void 0 ? [] : _c;
    var _d = (0, react_1.useReducer)(function (v) {
        return v + 1;
    }, 0), forceUpdate = _d[1];
    var _e = (0, react_1.useState)(''), inputValue = _e[0], setInputValue = _e[1];
    var _f = (0, react_1.useState)([]), list = _f[0], setList = _f[1];
    var _g = (0, react_1.useState)(0), activeIndex = _g[0], setActiveIndex = _g[1];
    var goTo = (0, react_1.useContext)(context_1.NavbarContext).goTo;
    // const source = useRef<CancelTokenSource>();
    var t = locales[lang];
    var tComponents = locale_1.default[lang];
    var latest = localStorage.getItem('arco-latest-search')
        ? JSON.parse(localStorage.getItem('arco-latest-search'))
        : [];
    var queryList = getQueryList(inputValue);
    var totalList = list.concat(queryList);
    var totalLength = totalList.length;
    function getExactList(inputValue) {
        var com = components_1.default
            .map(function (c) {
            return {
                label: tComponents[c],
                value: c,
            };
        })
            .filter(function (c) {
            return c.label.toLowerCase().indexOf(inputValue.toLowerCase()) > -1;
        })
            .map(function (c) {
            return {
                key: c.value,
                type: 'component',
                svg: react_1.default.createElement(react_1.default.Fragment, null, react_1.default.createElement('div', { className: list_module_less_1.default.n }, icon_2.default['Icon' + c.value + 'Gray']), react_1.default.createElement('div', { className: list_module_less_1.default.h }, icon_2.default['Icon' + c.value + 'Blue'])),
                backgroundColor: getBackgroundColor('Icon' + c.value + 'Gray'),
                text: highlightStr(c.label, inputValue),
                url: '/react' + (lang === 'en-US' ? '/en-US' : '') + '/components/' + (0, case_1.underscored)(c.value),
            };
        });
        var ecosystem = data_1.default
            .filter(function (d) {
            return d.keyword.find(function (dk) {
                return dk.toLowerCase().indexOf(inputValue.toLowerCase()) > -1;
            });
        })
            .map(function (d) {
            return __assign(__assign({}, d), {
                keyword: d.keyword.find(function (dk) {
                    return dk.toLowerCase().indexOf(inputValue.toLowerCase()) > -1;
                }),
            });
        })
            .map(function (d) {
            return __assign(__assign({}, d), {
                svg: icon_2.default[d.svg] || icon_2.default.IconBranding,
                backgroundColor: getBackgroundColor(d.svg),
                text: highlightStr(d.keyword, inputValue),
                borderRadius: 8,
            });
        });
        return com.concat(ecosystem);
    }
    var putInStorage = (0, react_1.useCallback)(function () {
        latest.unshift(__assign(__assign({}, totalList[activeIndex]), { inputValue: inputValue }));
        latest = latest.slice(0, 10);
        localStorage.setItem('arco-latest-search', JSON.stringify(latest));
    }, [activeIndex, inputValue, totalList]);
    var onKeyDown = (0, react_1.useCallback)(function (e) {
        var newActiveIndex = activeIndex;
        if (visible) {
            if (e.key === 'ArrowDown') {
                if (activeIndex + 1 > totalLength - 1) {
                    newActiveIndex = 0;
                }
                else {
                    newActiveIndex += 1;
                }
            }
            else if (e.key === 'ArrowUp') {
                if (activeIndex - 1 < 0) {
                    newActiveIndex = totalLength - 1;
                }
                else {
                    newActiveIndex -= 1;
                }
            }
            else if (e.key === 'Enter') {
                if (totalList[activeIndex] && totalList[activeIndex].url) {
                    goTo(totalList[activeIndex].url);
                    putInStorage();
                }
            }
        }
        setActiveIndex(newActiveIndex);
    }, [visible, activeIndex, totalLength, setActiveIndex, putInStorage, totalList]);
    (0, react_1.useEffect)(function () {
        window.addEventListener('keydown', onKeyDown);
        return function () {
            window.removeEventListener('keydown', onKeyDown);
        };
    }, [onKeyDown]);
    (0, react_1.useEffect)(function () {
        if (visible) {
            setInputValue('');
        }
    }, [visible]);
    function onChangeInputValue(e) {
        var iv = e.target.value;
        setActiveIndex(0);
        setInputValue(iv);
        setList(getExactList(iv));
        onChangeSearchInput && onChangeSearchInput(iv);
    }
    function getQueryList(inputValue) {
        return queryData
            .filter(function (q) {
            return q.description.toLowerCase().indexOf(inputValue.toLowerCase()) > -1;
        })
            .map(function (q) {
            var desc;
            var i = q.description.indexOf(inputValue);
            if (i > 50) {
                desc = q.description.slice(i - 50, i + 50);
            }
            else {
                desc = q.description.slice(0, 100);
            }
            return __assign(__assign({}, q), {
                title: q.title,
                description: highlightStr(desc, inputValue),
            });
        });
    }
    function onMouseEnterItem(index) {
        setActiveIndex(index);
    }
    function clearAllHistory() {
        localStorage.setItem('arco-latest-search', '[]');
        forceUpdate();
    }
    function onCloseHistory(e, key) {
        e.stopPropagation();
        localStorage.setItem('arco-latest-search', JSON.stringify(latest.filter(function (l) {
            return l.key !== key;
        })));
        forceUpdate();
    }
    function renderPreSearch() {
        return react_1.default.createElement('div', { className: style_module_less_1.default.content }, latest.length
            ? react_1.default.createElement('div', { className: history_module_less_1.default.history }, react_1.default.createElement('div', { className: history_module_less_1.default.header }, react_1.default.createElement('div', null, t.recentSearch), react_1.default.createElement(web_react_1.Link, { onClick: clearAllHistory }, t.clearAll)), react_1.default.createElement('ul', { className: history_module_less_1.default.list }, latest.map(function (l) {
                return react_1.default.createElement('li', {
                    className: history_module_less_1.default.item,
                    onClick: function () {
                        return goTo(l.url);
                    },
                    key: l.key,
                }, react_1.default.createElement(icon_1.IconHistory, null), react_1.default.createElement('span', { className: history_module_less_1.default.title }, l.inputValue), react_1.default.createElement('span', {
                    className: "".concat(history_module_less_1.default.close, " arco-icon-hover"),
                    onClick: function (e) {
                        return onCloseHistory(e, l.key);
                    },
                }, react_1.default.createElement(icon_1.IconClose, null)));
            })))
            : null, react_1.default.createElement('div', { className: hot_module_less_1.default.hot }, react_1.default.createElement('div', { className: hot_module_less_1.default.half }, react_1.default.createElement('div', { className: hot_module_less_1.default.title }, t.popularComponents), react_1.default.createElement('div', { className: hot_module_less_1.default.list }, popularComponents.map(function (p, index) {
            return react_1.default.createElement('div', {
                className: hot_module_less_1.default.item,
                onClick: function () {
                    return goTo(p.url);
                },
                key: index,
            }, react_1.default.createElement('div', { className: hot_module_less_1.default.svgWrapper }, p.svg), t[p.name], react_1.default.createElement('div', { className: hot_module_less_1.default.enterIcon }, icon_2.default.IconSearchIcEnter));
        }))), react_1.default.createElement('div', { className: hot_module_less_1.default.half }, react_1.default.createElement('div', { className: hot_module_less_1.default.title }, t.popularResource), react_1.default.createElement('div', { className: hot_module_less_1.default.list }, popularResource.map(function (p, index) {
            return react_1.default.createElement('div', {
                className: hot_module_less_1.default.item,
                onClick: function () {
                    return goTo(p.url, true);
                },
                key: index,
            }, react_1.default.createElement('div', {
                className: "".concat(hot_module_less_1.default.svgWrapper, " ").concat(hot_module_less_1.default.svgWrapperRect),
                style: { backgroundColor: p.color },
            }, p.svg), t[p.name], react_1.default.createElement('div', { className: hot_module_less_1.default.enterIcon }, icon_2.default.IconSearchIcEnter));
        })))));
    }
    return react_1.default.createElement(web_react_1.Modal, {
        alignCenter: false,
        style: { top: 150 },
        title: null,
        visible: visible,
        onOk: function () {
            return setVisible(false);
        },
        onCancel: function () {
            return setVisible(false);
        },
        footer: null,
        closable: false,
        className: style_module_less_1.default.modal,
        unmountOnExit: true,
    }, react_1.default.createElement('div', { className: style_module_less_1.default.inputWrapper }, react_1.default.createElement('div', { className: style_module_less_1.default.left }, react_1.default.createElement(icon_1.IconSearch, null), react_1.default.createElement('input', {
        value: inputValue,
        onChange: onChangeInputValue,
        placeholder: 'Search',
    })), react_1.default.createElement('div', { className: style_module_less_1.default.right }, react_1.default.createElement('div', { className: style_module_less_1.default.keyWorld }, t.openSearch), ' ', react_1.default.createElement('div', { className: style_module_less_2.default.command }, react_1.default.createElement('div', { className: style_module_less_2.default.key }, '\u2318'), react_1.default.createElement('div', { className: style_module_less_2.default.key }, 'K')), react_1.default.createElement('div', { className: style_module_less_1.default.keyWorld }, t.exit), react_1.default.createElement('div', { className: style_module_less_2.default.key, style: { width: 29 } }, 'ESC'))), inputValue
        ? react_1.default.createElement(react_1.default.Fragment, null, react_1.default.createElement('div', { className: style_module_less_1.default.count }, t.searchResult.replace('{0}', list.length + queryList.length)), list.length + queryList.length > 0
            ? react_1.default.createElement(list_1.default, {
                exactList: list,
                list: queryList,
                lang: lang,
                activeIndex: activeIndex,
                onMouseEnterItem: onMouseEnterItem,
                onClickItem: putInStorage,
            })
            : react_1.default.createElement(web_react_1.Empty, { style: { marginBottom: 32 } }))
        : renderPreSearch());
}
exports.default = SearchModal;
