// @m-design/m-site-navbar/lib/components/search-modal/list.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-nocheck
var react_1 = __importStar(require("react"));
var classnames_1 = __importDefault(require("classnames"));
var scroll_into_view_if_needed_1 = __importDefault(require("scroll-into-view-if-needed"));
var list_module_less_1 = __importDefault(require("./style/list.module.less"));
var context_1 = require("../../context");
var icon_1 = __importDefault(require("../../icon"));
var locales = {
    'zh-CN': {
        component: '组件',
        ecosystem: '生态',
    },
    'en-US': {
        component: 'Component',
        ecosystem: 'Ecosystem',
    },
};
function List(_a) {
    var _b = _a.exactList, exactList = _b === void 0 ? [] : _b, _c = _a.list, list = _c === void 0 ? [] : _c, _d = _a.lang, lang = _d === void 0 ? 'zh-CN' : _d, activeIndex = _a.activeIndex, onMouseEnterItem = _a.onMouseEnterItem, onClickItem = _a.onClickItem;
    var exactListRef = (0, react_1.useRef)(null);
    var queryListRef = (0, react_1.useRef)(null);
    var t = locales[lang];
    var goTo = (0, react_1.useContext)(context_1.NavbarContext).goTo;
    (0, react_1.useEffect)(function () {
        var allList = __spreadArray(__spreadArray([], exactListRef.current.childNodes, true), queryListRef.current.childNodes, true);
        if (allList.length) {
            (0, scroll_into_view_if_needed_1.default)(allList[activeIndex], {
                scrollMode: 'if-needed',
                block: 'nearest',
                inline: 'nearest',
            });
        }
    }, [activeIndex]);
    return react_1.default.createElement('div', { className: list_module_less_1.default.wrapper }, react_1.default.createElement('div', { className: list_module_less_1.default.exactList, ref: exactListRef }, exactList.map(function (ex, index) {
        var _a;
        var exactStyle = { backgroundColor: ex.backgroundColor };
        if (ex.borderRadius) {
            exactStyle.borderRadius = ex.borderRadius;
        }
        return react_1.default.createElement('div', {
            className: (0, classnames_1.default)(list_module_less_1.default.exactItem, ((_a = {}), (_a[list_module_less_1.default.itemActive] = activeIndex === index), _a)),
            key: ex.key,
            onMouseEnter: function () {
                return onMouseEnterItem(index);
            },
            onClick: function () {
                goTo(ex.url);
                onClickItem();
            },
        }, ex.svg &&
            react_1.default.createElement('div', { className: list_module_less_1.default.exactIcon, style: exactStyle }, ex.svg), react_1.default.createElement('div', {
            className: list_module_less_1.default.text,
            dangerouslySetInnerHTML: { __html: ex.text },
        }), react_1.default.createElement('div', { className: list_module_less_1.default.type }, t[ex.type]), react_1.default.createElement('div', { className: list_module_less_1.default.enterIcon }, icon_1.default.IconSearchIcEnter));
    })), react_1.default.createElement('div', { className: list_module_less_1.default.list, ref: queryListRef }, list.map(function (l, index) {
        var _a;
        return react_1.default.createElement('div', {
            className: (0, classnames_1.default)(list_module_less_1.default.listItem, ((_a = {}), (_a[list_module_less_1.default.itemActive] = activeIndex === index + exactList.length), _a)),
            onMouseEnter: function () {
                return onMouseEnterItem(exactList.length + index);
            },
            onClick: function () {
                goTo(l.url);
                onClickItem();
            },
            key: l.key,
        }, react_1.default.createElement('div', { className: list_module_less_1.default.listTitle }, l.title), react_1.default.createElement('div', {
            className: list_module_less_1.default.listDescription,
            dangerouslySetInnerHTML: { __html: l.description },
        }), react_1.default.createElement('div', { className: list_module_less_1.default.enterIcon }, icon_1.default.IconSearchIcEnter));
    })));
}
exports.default = List;
