// @arco-design/web-react/lib/_util/dayjs.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getResolvedDayjsLocaleName = exports.isDayjsArrayChange = exports.isDayjsChange = exports.getSortedDayjsArray = exports.getValueWithTime = exports.getDayjsValue = exports.getTimeFormat = exports.initializeDateLocale = exports.getNow = exports.methods = exports.dayjs = void 0;
var dayjs_1 = __importDefault(require("dayjs"));
var customParseFormat_1 = __importDefault(require("dayjs/plugin/customParseFormat"));
var isBetween_1 = __importDefault(require("dayjs/plugin/isBetween"));
var weekOfYear_1 = __importDefault(require("dayjs/plugin/weekOfYear"));
var advancedFormat_1 = __importDefault(require("dayjs/plugin/advancedFormat"));
var weekYear_1 = __importDefault(require("dayjs/plugin/weekYear"));
var quarterOfYear_1 = __importDefault(require("dayjs/plugin/quarterOfYear"));
require("dayjs/locale/zh-cn");
var is_1 = require("../_util/is");
var isMoment = (0, dayjs_1.default)()._isAMomentObject;
if (isMoment) {
    dayjs_1.default.extend = function () { };
}
// const methods = ['add', 'subtract', 'startOf', 'endOf', 'set'];
// methods.forEach((method) => {
//   const old = originDayjs.prototype[method];
//   originDayjs.prototype[`_${method}`] = function(...args) {
//     return isMoment ? old.call(this.clone(), ...args) : old(...args);
//   };
// });
var overwriteIsDayjs = function (_, Dayjs, dayjs) {
    dayjs = function (date, c) {
        if ((0, is_1.isDayjs)(date)) {
            return date.clone();
        }
        var cfg = typeof c === 'object' ? c : {};
        cfg.date = date;
        cfg.args = arguments; // eslint-disable-line prefer-rest-params
        return new Dayjs(cfg);
    };
    var proto = Dayjs.prototype;
    var old$Utils = proto.$utils;
    proto.$utils = function () {
        var newUtils = old$Utils();
        newUtils.i = is_1.isDayjs;
        return newUtils;
    };
    dayjs.isDayjs = is_1.isDayjs;
};
dayjs_1.default.extend(overwriteIsDayjs);
dayjs_1.default.extend(customParseFormat_1.default);
dayjs_1.default.extend(isBetween_1.default);
dayjs_1.default.extend(weekOfYear_1.default);
dayjs_1.default.extend(advancedFormat_1.default);
dayjs_1.default.extend(weekYear_1.default);
dayjs_1.default.extend(quarterOfYear_1.default);
exports.dayjs = dayjs_1.default;
// 兼容 moment
exports.methods = {
    add: function (time, value, unit) {
        return isMoment ? time.clone().add(value, unit) : time.add(value, unit);
    },
    subtract: function (time, value, unit) {
        return isMoment ? time.clone().subtract(value, unit) : time.subtract(value, unit);
    },
    startOf: function (time, unit) {
        return isMoment ? time.clone().startOf(unit) : time.startOf(unit);
    },
    endOf: function (time, unit) {
        return isMoment ? time.clone().endOf(unit) : time.endOf(unit);
    },
    set: function (time, unit, value) {
        return isMoment ? time.clone().set(unit, value) : time.set(unit, value);
    },
    isSameWeek: function (date1, date2, weekStart, localeName) {
        return isMoment
            ? date1.locale(localeName, { week: { dow: weekStart } }).isSame(date2, 'week')
            : date1.locale(__assign(__assign({}, exports.dayjs.Ls[localeName]), { weekStart: weekStart })).isSame(date2, 'week');
    },
};
function getNow() {
    return (0, exports.dayjs)();
}
exports.getNow = getNow;
function initializeDateLocale(localeName, weekStart) {
    if (isMoment) {
        exports.dayjs.updateLocale(localeName, { week: { dow: weekStart } });
    }
    else {
        exports.dayjs.locale(__assign(__assign({}, exports.dayjs.Ls[localeName]), { weekStart: weekStart }));
    }
}
exports.initializeDateLocale = initializeDateLocale;
function getTimeFormat(format) {
    var units = ['H', 'h', 'm', 's', 'A', 'a'];
    var timeFormat = '';
    units.some(function (unit) {
        if (format.indexOf(unit) !== -1) {
            timeFormat = "" + unit + format.split(" " + unit)[1];
            return true;
        }
        return false;
    });
    return timeFormat || 'HH:mm:ss';
}
exports.getTimeFormat = getTimeFormat;
function getDayjsValue(time, format) {
    if (!time) {
        return undefined;
    }
    var formatValue = function (value) {
        if ((0, is_1.isDayjs)(value)) {
            return (0, exports.dayjs)(value.valueOf());
        }
        if (typeof value === 'string') {
            return (0, exports.dayjs)(value, format);
        }
        return (0, exports.dayjs)(value);
    };
    if ((0, is_1.isArray)(time)) {
        return time.map(function (t) { return (t ? formatValue(t) : undefined); });
    }
    return formatValue(time);
}
exports.getDayjsValue = getDayjsValue;
function getValueWithTime(date, time) {
    if (date === void 0) { date = getNow(); }
    var y = date.year();
    var m = date.month();
    var d = date.date();
    if (time) {
        var returnTime = time;
        returnTime = exports.methods.set(returnTime, 'year', y);
        returnTime = exports.methods.set(returnTime, 'month', m);
        returnTime = exports.methods.set(returnTime, 'date', d);
        return returnTime;
    }
    return date;
}
exports.getValueWithTime = getValueWithTime;
function getSortedDayjsArray(values) {
    if (!values || !values[0] || !values[1]) {
        return values;
    }
    var newValues = __spreadArray([], __read(values), false);
    newValues.sort(function (a, b) { return a.valueOf() - b.valueOf(); });
    return newValues;
}
exports.getSortedDayjsArray = getSortedDayjsArray;
function isDayjsChange(prevValue, currentValue) {
    if (currentValue === undefined && prevValue === undefined) {
        return false;
    }
    return ((currentValue && !prevValue) ||
        (!currentValue && prevValue) ||
        (0, exports.dayjs)(currentValue).valueOf() !== (0, exports.dayjs)(prevValue).valueOf());
}
exports.isDayjsChange = isDayjsChange;
function isDayjsArrayChange(prevValue, currentValue) {
    if (currentValue === undefined && prevValue === undefined) {
        return false;
    }
    return ((currentValue && !prevValue) ||
        (!currentValue && prevValue) ||
        ((0, is_1.isArray)(currentValue) &&
            (0, is_1.isArray)(prevValue) &&
            (0, exports.dayjs)(currentValue[0]).valueOf() !== (0, exports.dayjs)(prevValue[0]).valueOf()) ||
        (0, exports.dayjs)(currentValue[1]).valueOf() !== (0, exports.dayjs)(prevValue[1]).valueOf());
}
exports.isDayjsArrayChange = isDayjsArrayChange;
function getResolvedDayjsLocaleName(localeName) {
    var locale = localeName.toLocaleLowerCase();
    return locale === 'en-us' ? 'en' : locale;
}
exports.getResolvedDayjsLocaleName = getResolvedDayjsLocaleName;
