// @m-design/m-site-navbar/lib/components/panel/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-nocheck
var react_1 = __importStar(require("react"));
var web_react_1 = require("@arco-design/web-react");
var icon_1 = require("@arco-design/web-react/icon");
var classnames_1 = __importDefault(require("classnames"));
var gsap_1 = __importDefault(require("gsap"));
var icon_2 = __importDefault(require("../../icon"));
var logo_m_svg_1 = __importDefault(require("../../assets/logo-m.svg"));
var index_module_less_1 = __importDefault(require("./index.module.less"));
var locales = {
    "zh-CN": {
        pro: "中后台最佳实践",
        material: "物料平台",
        iconbox: "图标平台",
        designLab: "设计系统配置平台",
        chartSpace: "图表库",
        brand: "品牌资源平台",
        fontMall: "字体配置平台",
        toBeOpen: "即将开放",
        palette: "色彩配置工具",
    },
    "en-US": {
        pro: "Best Practices",
        material: "Material Market",
        iconbox: "Icon Platform",
        designLab: "Design System Lab",
        chartSpace: "Arco Chart Space",
        brand: "BrandStore",
        fontMall: "FontMall",
        toBeOpen: "Open soon",
        palette: "Color Configuration Tool",
    },
};
function Popup(_a) {
    var logoData = _a.logoData, t = _a.t, isHome = _a.isHome, isMobile = _a.isMobile;
    return (react_1.default.createElement("div", { className: (0, classnames_1.default)(index_module_less_1.default.wrapper, isMobile ? index_module_less_1.default.m : "") },
        !isHome && (react_1.default.createElement("a", { className: index_module_less_1.default.home, href: "/" },
            react_1.default.createElement(logo_m_svg_1.default, { width: "143px", height: "24px" }),
            react_1.default.createElement(icon_1.IconArrowRight, null))),
        react_1.default.createElement("div", { className: index_module_less_1.default.panel }, logoData.map(function (l, index) { return (react_1.default.createElement("a", { 
            // eslint-disable-next-line react/no-array-index-key
            key: index, className: index_module_less_1.default.item, href: l.url, onClick: function () {
                if (!l.url) {
                    web_react_1.Message.info(t.toBeOpen);
                }
            } },
            react_1.default.createElement("div", { className: index_module_less_1.default.logo },
                react_1.default.createElement("span", { className: index_module_less_1.default.h }, l.logo),
                react_1.default.createElement("span", { className: index_module_less_1.default.n }, l.logoB)),
            react_1.default.createElement("div", { className: index_module_less_1.default.text },
                react_1.default.createElement("p", { className: index_module_less_1.default.title }, l.title),
                react_1.default.createElement("span", { className: index_module_less_1.default.desc }, l.des)))); }))));
}
function Panel(_a) {
    var lang = _a.lang, getPopupContainer = _a.getPopupContainer, isHome = _a.isHome, isMobile = _a.isMobile;
    var _b = (0, react_1.useState)(false), visible = _b[0], setVisible = _b[1];
    var t = locales[lang];
    var logoData = [
        {
            logo: icon_2.default.IconLogoArcoPro,
            logoB: icon_2.default.IconLogoArcoProBlue,
            title: "M Design Pro",
            des: t.pro,
            url: "/pro",
        },
        {
            logo: icon_2.default.IconLogoDesignLab,
            logoB: icon_2.default.IconLogoDesignLabBlue,
            title: "DesignLab",
            des: t.designLab,
            url: "/themes",
        },
        {
            logo: icon_2.default.IconLogoArcoMaterial,
            logoB: icon_2.default.IconLogoArcoMaterialBlue,
            title: "MaterialMarket",
            des: t.material,
            url: "/material",
        },
        // {
        //   logo: icon.IconLogoIconBox,
        //   logoB: icon.IconLogoIconBoxBlue,
        //   title: 'IconBox',
        //   des: t.iconbox,
        //   url: '/iconbox',
        // },
    ];
    (0, react_1.useEffect)(function () {
        var panelTL = gsap_1.default.timeline();
        var panelTL2 = gsap_1.default.timeline();
        if (visible) {
            panelTL.fromTo(".".concat(index_module_less_1.default.item), {
                opacity: 0,
                duration: 0.3,
                y: -4,
                delay: 0.1,
                stagger: {
                    amount: 0.2,
                },
            }, {
                opacity: 1,
                duration: 0.3,
                y: 0,
                delay: 0.1,
                stagger: {
                    amount: 0.2,
                },
            });
        }
        else {
            panelTL2.to(".".concat(index_module_less_1.default.item), {
                opacity: 0,
                duration: 0.2,
                y: -4,
            });
        }
    }, [visible]);
    var triggerProps = {
        popupVisible: visible,
        onVisibleChange: function (v) {
            return setVisible(v);
        },
        trigger: "hover",
        popup: function () {
            return react_1.default.createElement(Popup, {
                logoData: logoData,
                t: t,
                isHome: isHome,
                isMobile: isMobile,
            });
        },
        position: "bl",
        popupAlign: { bottom: [12, 4] },
        duration: 600,
        classNames: "down",
        getPopupContainer: getPopupContainer,
    };
    return (
    // eslint-disable-next-line react/jsx-props-no-spreading
    react_1.default.createElement(web_react_1.Trigger, __assign({}, triggerProps),
        react_1.default.createElement("div", { className: (0, classnames_1.default)(index_module_less_1.default.icon, isMobile ? index_module_less_1.default.mobile : "") },
            react_1.default.createElement("div", { className: index_module_less_1.default.iconInner }, react_1.default.createElement("svg", {
                width: "20",
                height: "20",
                viewBox: "0 0 20 20",
                fill: "none",
                xmlns: "http://www.w3.org/2000/svg",
            }, react_1.default.createElement("path", {
                fillRule: "evenodd",
                clipRule: "evenodd",
                d: "M2.5 5L5 5L5 2.5L2.5 2.5L2.5 5ZM5 11.25L2.5 11.25L2.5 8.75L5 8.75L5 11.25ZM17.5 11.25L15 11.25L15 8.75L17.5 8.75L17.5 11.25ZM8.75 11.25L11.25 11.25L11.25 8.75L8.75 8.75L8.75 11.25ZM5 17.5L2.5 17.5L2.5 15L5 15L5 17.5ZM15 17.5L17.5 17.5L17.5 15L15 15L15 17.5ZM11.25 17.5L8.75 17.5L8.75 15L11.25 15L11.25 17.5ZM17.5 5L15 5L15 2.5L17.5 2.5L17.5 5ZM8.75 5L11.25 5L11.25 2.5L8.75 2.5L8.75 5Z",
                fill: "currentColor",
            }))))));
}
exports.default = Panel;
