// @arco-design/web-react/lib/Typography/base.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var useIsomorphicLayoutEffect_1 = __importDefault(require("../_util/hooks/useIsomorphicLayoutEffect"));
var ConfigProvider_1 = require("../ConfigProvider");
var operations_1 = __importDefault(require("./operations"));
var classNames_1 = __importDefault(require("../_util/classNames"));
var edit_content_1 = __importDefault(require("./edit-content"));
var is_1 = require("../_util/is");
var useResizeObserver_1 = __importDefault(require("../_util/hooks/useResizeObserver"));
var utils_1 = require("./utils");
var Tooltip_1 = __importDefault(require("../Tooltip"));
var Popover_1 = __importDefault(require("../Popover"));
var raf_1 = require("../_util/raf");
var omit_1 = __importDefault(require("../_util/omit"));
var useUpdate_1 = __importDefault(require("../_util/hooks/useUpdate"));
var mergedToString_1 = __importDefault(require("../_util/mergedToString"));
function getClassNameAndComponentName(props, prefixCls) {
    var type = props.type, bold = props.bold, disabled = props.disabled, mark = props.mark, underline = props.underline, propDelete = props.delete, code = props.code;
    var component = [];
    var className = [];
    if (type) {
        className.push(prefixCls + "-" + type);
    }
    if (disabled) {
        className.push(prefixCls + "-disabled");
    }
    if (bold) {
        component.push('b');
    }
    if (underline) {
        component.push('u');
    }
    if (propDelete) {
        component.push('del');
    }
    if (code) {
        component.push('code');
    }
    if (mark) {
        component.push('mark');
    }
    return {
        component: component,
        className: className,
    };
}
function wrap(content, component, props) {
    var currentContent = content;
    component.forEach(function (c) {
        var _props = (0, is_1.isObject)(props.mark) && props.mark.color
            ? { style: { backgroundColor: props.mark.color } }
            : {};
        currentContent = react_1.default.createElement(c, _props, currentContent);
    });
    return currentContent;
}
function Base(props) {
    var componentType = props.componentType, style = props.style, className = props.className, children = props.children, editable = props.editable, ellipsis = props.ellipsis, heading = props.heading, blockquote = props.blockquote, rest = __rest(props, ["componentType", "style", "className", "children", "editable", "ellipsis", "heading", "blockquote"]);
    var getPrefixCls = (0, react_1.useContext)(ConfigProvider_1.ConfigContext).getPrefixCls;
    var prefixCls = getPrefixCls('typography');
    var rafId = (0, react_1.useRef)();
    var _a = getClassNameAndComponentName(props, prefixCls), component = _a.component, componentClassName = _a.className;
    var _b = __read((0, react_1.useState)(false), 2), editing = _b[0], setEditing = _b[1];
    var _c = __read((0, react_1.useState)(false), 2), isEllipsis = _c[0], setEllipsis = _c[1];
    var _d = __read((0, react_1.useState)(false), 2), expanding = _d[0], setExpanding = _d[1];
    var _e = __read((0, react_1.useState)(''), 2), ellipsisText = _e[0], setEllipsisText = _e[1];
    var _f = __read((0, react_1.useState)(false), 2), measuring = _f[0], setMeasuring = _f[1];
    var componentRef = (0, react_1.useRef)(null);
    var editableConfig = (0, is_1.isObject)(editable) ? editable : {};
    var mergedEditing = 'editing' in editableConfig ? editableConfig.editing : editing;
    var ellipsisConfig = ellipsis
        ? __assign({ rows: 1, ellipsisStr: '...' }, ((0, is_1.isObject)(ellipsis) ? ellipsis : {})) : {};
    function renderOperations(forceShowExpand) {
        return (react_1.default.createElement(react_1.default.Fragment, null,
            react_1.default.createElement(operations_1.default, __assign({}, props, { setEditing: setEditing, onClickExpand: onClickExpand, expanding: expanding, isEllipsis: isEllipsis, forceShowExpand: forceShowExpand }))));
    }
    function onClickExpand() {
        setExpanding(!expanding);
        ellipsisConfig.onExpand && ellipsisConfig.onExpand(!expanding);
    }
    var resizeOnNextFrame = function () {
        (0, raf_1.caf)(rafId.current);
        rafId.current = (0, raf_1.raf)(function () {
            calcEllipsis();
        });
    };
    var _g = (0, useResizeObserver_1.default)(resizeOnNextFrame), cor = _g.cor, dor = _g.dor, currentOr = _g.currentOr;
    (0, useIsomorphicLayoutEffect_1.default)(function () {
        if (!currentOr) {
            resizeOnNextFrame();
        }
    }, [currentOr]);
    (0, useUpdate_1.default)(function () {
        ellipsisConfig.onEllipsis && ellipsisConfig.onEllipsis(isEllipsis);
    }, [isEllipsis]);
    (0, useIsomorphicLayoutEffect_1.default)(function () {
        if (componentRef.current) {
            cor(componentRef.current);
        }
        return function () {
            dor();
            (0, raf_1.caf)(rafId.current);
        };
    }, [
        children,
        expanding,
        isEllipsis,
        editing,
        ellipsisConfig.suffix,
        ellipsisConfig.ellipsisStr,
        ellipsisConfig.expandable,
        ellipsisConfig.expandNodes,
        ellipsisConfig.rows,
    ]);
    function calcEllipsis() {
        var currentEllipsis = isEllipsis;
        if (editing) {
            return;
        }
        if (ellipsisConfig.rows) {
            // In ellipsis mode, if the user manually expands, there is no need to calculate ellipsis and ellipsisText;
            if (expanding) {
                return;
            }
            setMeasuring(true);
            var _a = (0, utils_1.measure)(componentRef.current, ellipsisConfig, renderOperations(!!ellipsisConfig.expandable), children), ellipsis_1 = _a.ellipsis, text = _a.text;
            setMeasuring(false);
            if (ellipsis_1 && text) {
                setEllipsisText(text);
            }
            if (ellipsis_1 !== currentEllipsis) {
                setEllipsis(ellipsis_1);
            }
        }
        else {
            var isEllipsis_1 = !!ellipsisConfig.rows;
            if (isEllipsis_1 !== currentEllipsis) {
                setEllipsis(isEllipsis_1);
            }
        }
    }
    function renderContent() {
        var fullText = (0, mergedToString_1.default)(react_1.default.Children.toArray(children));
        var ellipsisStr = ellipsisConfig.ellipsisStr !== undefined ? ellipsisConfig.ellipsisStr : '...';
        var suffix = ellipsisConfig.suffix !== undefined && ellipsisConfig.suffix;
        var showTooltip = ellipsisConfig.showTooltip;
        var tooltipType = (0, is_1.isObject)(ellipsisConfig.showTooltip)
            ? ellipsisConfig.showTooltip.type === 'popover'
                ? 'popover'
                : 'tooltip'
            : 'tooltip';
        var tooltipProps = (0, is_1.isObject)(ellipsisConfig.showTooltip)
            ? ellipsisConfig.showTooltip.props || {}
            : {};
        var TooltipComponent = (tooltipType === 'popover' ? Popover_1.default : Tooltip_1.default);
        var titleProps = isEllipsis && !showTooltip && !expanding ? { title: fullText } : {};
        var baseProps = __assign({ ref: componentRef, style: style }, titleProps);
        function renderInnerContent() {
            var text = isEllipsis && !expanding ? ellipsisText : children;
            var innerText = component.length ? wrap(text, component, props) : text;
            return (react_1.default.createElement(react_1.default.Fragment, null,
                isEllipsis && showTooltip && !expanding ? (react_1.default.createElement(TooltipComponent, __assign({ content: fullText }, tooltipProps), innerText)) : (innerText),
                measuring || (isEllipsis && !expanding) ? ellipsisStr : null,
                suffix,
                renderOperations(measuring ? !!ellipsisConfig.expandable : undefined)));
        }
        var TextComponent;
        if (componentType === 'Paragraph') {
            TextComponent = blockquote ? 'blockquote' : 'div';
        }
        else if (componentType === 'Title') {
            TextComponent = "h" + heading;
        }
        else if (componentType === 'Text') {
            TextComponent = ellipsis ? 'div' : 'span';
        }
        return (react_1.default.createElement(TextComponent, __assign({ className: (0, classNames_1.default)(prefixCls, componentClassName, className) }, baseProps, (0, omit_1.default)(rest, [
            'spacing',
            'type',
            'close',
            'bold',
            'disabled',
            'mark',
            'underline',
            'delete',
            'code',
            'copyable',
            'isEllipsis',
            'expanding',
            'onClickExpand',
            'setEditing',
            'forceShowExpand',
        ])), renderInnerContent()));
    }
    return mergedEditing ? (react_1.default.createElement(edit_content_1.default, __assign({}, props, { prefixCls: prefixCls, setEditing: setEditing, editableConfig: editableConfig }))) : (renderContent());
}
exports.default = Base;
