// @m-design/m-site-navbar/es/components/select/index.js
// @ts-nocheck
import React from 'react';
import { Button, Select } from '@arco-design/web-react';
import { IconCheck } from '@arco-design/web-react/icon';
import styles from './style.module.less';
function NavSelect(_a) {
    var value = _a.value, defaultValue = _a.defaultValue, options = _a.options, onChange = _a.onChange, triggerElement = _a.triggerElement, button = _a.button, buttonValue = _a.buttonValue, buttonIcon = _a.buttonIcon, onClickButton = _a.onClickButton;
    var selectProps = button
        ? {
            options: options.map(function (o) {
                var option = o;
                return {
                    label: React.createElement('span', null, React.createElement(IconCheck, {
                        style: {
                            marginRight: 10,
                            fontSize: 12,
                            opacity: buttonValue === option.value ? 1 : 0,
                        },
                    }), option.label),
                    value: option.value,
                };
            }),
            triggerElement: React.createElement(Button, {
                className: styles.btnIcon,
                type: 'text',
                icon: buttonIcon,
                size: 'mini',
                onClick: onClickButton,
            }),
        }
        : {};
    if (value) {
        selectProps.value = value;
    }
    else {
        selectProps.defaultValue = defaultValue;
    }
    return React.createElement(Select, Object.assign({
        trigger: 'hover',
        options: options,
        bordered: false,
        onChange: onChange,
        triggerProps: {
            autoAlignPopupWidth: false,
            autoAlignPopupMinWidth: true,
        },
        className: styles.select,
        triggerElement: triggerElement,
    }, selectProps));
}
export default NavSelect;
