// @m-design/m-site-footer/es/index.js
// @ts-nocheck
import React, { useState, useRef } from 'react';
import { IconDown } from '@arco-design/web-react/icon';
import cs from 'classnames';
// import ResizeObserver from './ResizeObserver';
import ResizeObserver from 'rc-resize-observer';
// import Logo from './ArcoLogo';
import styles from './style/footer.module.less';
var locales = {
    'zh-CN': {
        design: '设计',
        component: '组件',
        ecosystem: '生态产品',
        resource: '资源',
        feedback: '反馈',
        about: '关于 Arco',
        overview: '组件索引',
        spec: '设计规范',
        principle: '设计原则',
        start: '快速开始',
        question: '常见问题',
        migrate: '从 v1 到 v2',
        changelog: '更新日志',
        pro: '中后台最佳实践',
        componentFigma: '组件库 Figma 资源',
        iconFigma: '图标 Figma 资源',
        toolWebpack: 'Webpack Plugin',
        designLab: '设计系统配置平台',
        material: '物料平台',
    },
};
function Footer(props) {
    var style = props.style, logo = props.logo, lang = props.lang, isDoc = props.isDoc;
    var t = locales[lang];
    var footerRef = useRef(null);
    var _b = useState(''), screen = _b[0], setScreen = _b[1];
    var _c = useState(''), expandKey = _c[0], setExpandKey = _c[1];
    function updateScreen() {
        var width = footerRef.current.clientWidth;
        if (width >= 1600) {
            setScreen('xxl');
        }
        else if (width >= 1200 && width < 1600) {
            setScreen('xl');
        }
        else if (width >= 992 && width < 1200) {
            setScreen('lg');
        }
        else if (width >= 768 && width < 992) {
            setScreen('md');
        }
        else if (width < 768) {
            setScreen('mobile');
        }
    }
    function handleExpand(key) {
        if (screen === 'mobile') {
            if (expandKey === key) {
                setExpandKey('');
            }
            else {
                setExpandKey(key);
            }
        }
    }
    function getListHeightStyle(key, count) {
        return screen === 'mobile' ? { height: expandKey === key ? 48 * count : 0 } : {};
    }
    function getOpenClassName(key) {
        var _a = {};
        _a["".concat(styles.open)] = key === expandKey;
        return cs(styles.title, _a);
    }
    return (React.createElement(ResizeObserver, { onResize: updateScreen },
        React.createElement("footer", { ref: footerRef, className: cs(styles.footer, styles[screen], isDoc ? styles.document : ''), style: style },
            React.createElement("div", { className: styles.footerInner },
                React.createElement("div", { className: styles.content },
                    React.createElement("div", { className: styles.logo }, logo),
                    React.createElement("div", { className: styles.lw },
                        React.createElement("div", { className: styles.li },
                            React.createElement("div", { className: getOpenClassName('design'), onClick: function () { return handleExpand('design'); } },
                                t.design,
                                React.createElement(IconDown, null)),
                            React.createElement("div", { className: styles.list, style: getListHeightStyle('design', 2) },
                                React.createElement("a", { className: styles.item, href: "/docs/spec/introduce" }, t.spec),
                                React.createElement("a", { className: styles.item, href: "/docs/spec/values-of-arcodesign" }, "\u8BBE\u8BA1\u4EF7\u503C\u89C2"))),
                        React.createElement("div", { className: styles.li },
                            React.createElement("div", { className: getOpenClassName('component'), onClick: function () { return handleExpand('component'); } },
                                t.component,
                                React.createElement(IconDown, null)),
                            React.createElement("div", { className: styles.list, style: getListHeightStyle('component', 5) },
                                React.createElement("a", { className: styles.item, href: "/react/docs/start" }, t.start),
                                React.createElement("a", { className: styles.item, href: "/react/docs/question" }, t.question),
                                React.createElement("a", { className: styles.item, href: "/react/docs/changelog" }, t.changelog))),
                        React.createElement("div", { className: cs(styles.li, styles.liwd) },
                            React.createElement("div", { className: getOpenClassName('ecosystem'), onClick: function () { return handleExpand('ecosystem'); } },
                                t.ecosystem,
                                React.createElement(IconDown, null)),
                            React.createElement("div", { className: styles.list, style: getListHeightStyle('ecosystem', 7) },
                                React.createElement("a", { className: styles.item, href: "/themes/stores" }, t.designLab),
                                React.createElement("a", { className: styles.item, href: "/material" }, t.material))),
                        React.createElement("div", { className: styles.li },
                            React.createElement("div", { className: getOpenClassName('resource'), onClick: function () { return handleExpand('resource'); } },
                                t.resource,
                                React.createElement(IconDown, null)),
                            React.createElement("div", { className: styles.list, style: getListHeightStyle('resource', 6) },
                                React.createElement("a", { className: styles.item, href: "https://github.com/arco-design/arco-webpack-plugin" }, t.toolWebpack))))),
                React.createElement("div", { className: styles.bottom },
                    React.createElement("div", { className: styles.left },
                        React.createElement("div", { className: styles.copyright },
                            React.createElement("span", null, "UED & \u524D\u7AEF "),
                            React.createElement("span", null, "Copyright Meetsocial 2021"))))))));
}
Footer.defaultProps = {
    logo: React.createElement("span", null, "M Design"),
    lang: 'zh-CN',
};
export default Footer;
