// @arco-design/web-react/lib/Cascader/cascader.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultFieldNames = void 0;
var isEqualWith_1 = __importDefault(require("lodash/isEqualWith"));
var react_1 = __importStar(require("react"));
var is_1 = require("../_util/is");
var Trigger_1 = __importDefault(require("../Trigger"));
var list_1 = __importDefault(require("./panel/list"));
var search_panel_1 = __importDefault(require("./panel/search-panel"));
var ConfigProvider_1 = require("../ConfigProvider");
var store_1 = __importDefault(require("./base/store"));
var select_view_1 = __importDefault(require("../_class/select-view"));
var classNames_1 = __importDefault(require("../_util/classNames"));
var useMergeValue_1 = __importDefault(require("../_util/hooks/useMergeValue"));
var useUpdate_1 = __importDefault(require("../_util/hooks/useUpdate"));
var keycode_1 = require("../_util/keycode");
var useRefCurrent_1 = __importDefault(require("./hook/useRefCurrent"));
var useMergeProps_1 = __importDefault(require("../_util/hooks/useMergeProps"));
function isEmptyValue(value) {
    return !value || ((0, is_1.isArray)(value) && value.length === 0);
}
exports.DefaultFieldNames = {
    label: 'label',
    value: 'value',
    isLeaf: 'isLeaf',
    children: 'children',
    disabled: 'disabled',
};
function getConfig(props) {
    return {
        showEmptyChildren: props.showEmptyChildren,
        changeOnSelect: props.changeOnSelect,
        lazyload: !!props.loadMore,
        fieldNames: props.fieldNames,
        filterOption: props.filterOption,
    };
}
function getStore(props, value) {
    var tmp = value ? (Array.isArray(value[0]) ? value : [value]) : [];
    return new store_1.default(props.options || [], tmp, getConfig(props));
}
var formatValue = function (value, isMultiple) {
    if (value === undefined) {
        return [];
    }
    if (isMultiple) {
        return value;
    }
    return [value];
};
var defaultProps = {
    options: [],
    bordered: true,
    fieldNames: exports.DefaultFieldNames,
    trigger: 'click',
    expandTrigger: 'click',
};
function Cascader(baseProps, ref) {
    var _a = (0, react_1.useContext)(ConfigProvider_1.ConfigContext), getPrefixCls = _a.getPrefixCls, renderEmpty = _a.renderEmpty, componentConfig = _a.componentConfig;
    var props = (0, useMergeProps_1.default)(baseProps, defaultProps, componentConfig === null || componentConfig === void 0 ? void 0 : componentConfig.Cascader);
    var disabled = props.disabled, renderFormat = props.renderFormat, getPopupContainer = props.getPopupContainer, children = props.children, triggerProps = props.triggerProps, expandTrigger = props.expandTrigger;
    var prefixCls = getPrefixCls('cascader');
    var isMultiple = props.mode === 'multiple';
    var timerRef = (0, react_1.useRef)(null);
    var _b = __read((0, react_1.useState)(''), 2), inputValue = _b[0], setInputValue = _b[1];
    // 暂存被选中的值对应的节点。仅在onSearch的时候用到
    // 避免出现下拉列表改变，之前选中的option找不到对应的节点，展示上会出问题。
    var stashNodes = (0, react_1.useRef)([]);
    var propsValue = 'value' in props ? formatValue(props.value, isMultiple) : undefined;
    // 为什么不用useMergeValue呢？ 因为在props.value改变时，需要进行一些选中状态的处理，见以下 useUpdate
    var _c = __read((0, react_1.useState)(propsValue || ('defaultValue' in props ? formatValue(props.defaultValue, isMultiple) : [])), 2), stateValue = _c[0], setValue = _c[1];
    var mergeValue = 'value' in props ? propsValue : stateValue;
    var _d = __read((0, useMergeValue_1.default)(false, {
        value: props.popupVisible,
        defaultValue: props.defaultPopupVisible,
    }), 2), popupVisible = _d[0], setPopupVisible = _d[1];
    var selectRef = (0, react_1.useRef)(null);
    var store = (0, useRefCurrent_1.default)(function () { return getStore(props, mergeValue); }, [
        JSON.stringify(getConfig(props)),
        props.options,
    ]);
    (0, react_1.useEffect)(function () {
        var clearTimer = function () {
            clearTimeout(timerRef.current);
            timerRef.current = null;
        };
        if (!popupVisible && inputValue) {
            if (timerRef.current) {
                clearTimer();
            }
            timerRef.current = setTimeout(function () {
                setInputValue('');
                timerRef.current = null;
            }, 200);
        }
        return function () {
            clearTimer();
        };
    }, [popupVisible]);
    (0, useUpdate_1.default)(function () {
        setValue(function (stateValue) {
            var newValue = formatValue(props.value, isMultiple);
            if (!(0, isEqualWith_1.default)(stateValue, newValue)) {
                store.setNodeCheckedByValue(newValue);
                return newValue;
            }
            return stateValue;
        });
    }, [props.value, isMultiple]);
    (0, react_1.useImperativeHandle)(ref, function () { return selectRef.current; }, []);
    var updateStashNodes = function (nodes) {
        stashNodes.current = [].concat(nodes, stashNodes.current);
    };
    var getSelectedOptionsByValue = function (values) {
        var nodes = store.getCheckedNodes().concat(stashNodes.current);
        var result = [];
        values.map(function (value) {
            var node = nodes.find(function (item) { return (0, isEqualWith_1.default)(item.pathValue, value); });
            if (node) {
                result.push(node.getPathNodes().map(function (x) { return x._data; }));
            }
        });
        return result;
    };
    var handleVisibleChange = (0, react_1.useCallback)(function (newVisible) {
        if (newVisible !== popupVisible) {
            props.onVisibleChange && props.onVisibleChange(newVisible);
            if (!('popupVisible' in props)) {
                setPopupVisible(newVisible);
            }
        }
    }, [props.onVisibleChange, popupVisible]);
    var renderText = (0, react_1.useCallback)(function (value) {
        var _a;
        // store 中不存在时，从stashNodes.current中找一下对应节点
        var options = getSelectedOptionsByValue([value])[0] || [];
        var text;
        var valueShow = (0, is_1.isArray)(value) ? value.map(function (x) { return String(x); }) : [];
        if (options.length) {
            valueShow = options.map(function (x) { return x.label; });
        }
        if ((0, is_1.isFunction)(renderFormat)) {
            text = renderFormat(valueShow);
        }
        else if (valueShow.every(function (v) { return (0, is_1.isString)(v); })) {
            text = valueShow.join(' / ');
        }
        else {
            text = valueShow.reduce(function (total, item, index) {
                return total.concat(index === 0 ? [item] : [' / ', item]);
            }, []);
        }
        return {
            text: text || '',
            disabled: (_a = options[options.length - 1]) === null || _a === void 0 ? void 0 : _a.disabled,
        };
    }, [store, renderFormat]);
    // isTouch: 是否是通过点击remove图标或者清除图标改变的值
    var handleChange = function (newValue, isTouch) {
        setValue(function (mergeValue) {
            var _a;
            var onChange = props.onChange, changeOnSelect = props.changeOnSelect, expandTrigger = props.expandTrigger;
            var isSame = (0, isEqualWith_1.default)(mergeValue, newValue);
            if (!isSame) {
                if (isTouch || !isMultiple) {
                    store.setNodeCheckedByValue(newValue);
                }
            }
            var nodes = store.getCheckedNodes();
            !isSame && updateStashNodes(nodes);
            var selectedOptions = getSelectedOptionsByValue(newValue);
            if (!isSame) {
                var _value = isMultiple ? newValue : newValue[0];
                var _selectedOptions = isMultiple ? selectedOptions : selectedOptions[0];
                onChange &&
                    onChange(_value, _selectedOptions, {
                        dropdownVisible: popupVisible,
                    });
            }
            if (!isMultiple) {
                if (inputValue) {
                    // 单选时选择搜索项，直接关闭面板
                    handleVisibleChange(false);
                }
                else if ((selectedOptions[0] && ((_a = selectedOptions[0][selectedOptions[0].length - 1]) === null || _a === void 0 ? void 0 : _a.isLeaf)) ||
                    (changeOnSelect && expandTrigger === 'hover')) {
                    handleVisibleChange(false);
                }
            }
            // 这里直接通过setValue修改stateValue是为了节省受控模式下，不断通过外部value查找节点，计算选中状态的操作。
            // 和useUpdate配合，在statevalue和外部传入的value不相等的时候才进行计算。
            return isSame ? mergeValue : newValue;
        });
    };
    var onRemoveCheckedItem = function (item, index, e) {
        e.stopPropagation();
        if (item.disabled) {
            return;
        }
        var newValue = mergeValue.filter(function (_, i) { return i !== index; });
        handleChange(newValue, true);
    };
    var renderEmptyEle = function (width) {
        var wd = width || (selectRef.current && selectRef.current.getWidth());
        return (react_1.default.createElement("div", { className: prefixCls + "-list-empty", style: { width: wd } }, props.notFoundContent || renderEmpty('Cascader')));
    };
    var renderPopup = function () {
        var _a;
        var width = selectRef.current && selectRef.current.getWidth();
        var showSearchPanel = !(0, is_1.isFunction)(props.onSearch) && !!inputValue;
        var dropdownRender = (0, is_1.isFunction)(props.dropdownRender) ? props.dropdownRender : function (menu) { return menu; };
        return (react_1.default.createElement("div", { className: (0, classNames_1.default)(prefixCls + "-popup", (_a = {},
                _a[prefixCls + "-popup-trigger-hover"] = props.expandTrigger === 'hover',
                _a)) }, dropdownRender(react_1.default.createElement("div", { className: prefixCls + "-popup-inner", onMouseDown: function (e) { return e.preventDefault(); } }, showSearchPanel ? (react_1.default.createElement(search_panel_1.default, { style: { minWidth: width }, store: store, inputValue: inputValue, renderEmpty: function () { return renderEmptyEle(width); }, multiple: isMultiple, onChange: handleChange, prefixCls: prefixCls, onEsc: function () {
                handleVisibleChange(false);
            }, value: mergeValue })) : (react_1.default.createElement(list_1.default, { expandTrigger: expandTrigger, store: store, dropdownColumnRender: props.dropdownColumnRender, renderOption: props.renderOption, changeOnSelect: props.changeOnSelect, showEmptyChildren: props.showEmptyChildren || !!props.loadMore, multiple: isMultiple, onChange: handleChange, loadMore: props.loadMore, prefixCls: prefixCls, renderEmpty: renderEmptyEle, popupVisible: popupVisible, value: mergeValue, renderFooter: props.renderFooter, onEsc: function () {
                handleVisibleChange(false);
            }, onDoubleClickOption: function () {
                if (props.changeOnSelect && !isMultiple) {
                    handleVisibleChange(false);
                }
            } }))))));
    };
    return (react_1.default.createElement(Trigger_1.default, __assign({ popup: renderPopup, trigger: props.trigger, disabled: disabled, getPopupContainer: getPopupContainer, position: "bl", classNames: "slideDynamicOrigin", popupAlign: { bottom: 4 }, 
        // 动态加载时，unmountOnExit 默认为false。
        unmountOnExit: 'unmountOnExit' in props ? props.unmountOnExit : !(0, is_1.isFunction)(props.loadMore), popupVisible: popupVisible }, triggerProps, { onVisibleChange: handleVisibleChange }), children || (react_1.default.createElement(select_view_1.default, __assign({}, props, { ref: selectRef, popupVisible: popupVisible, value: isMultiple ? mergeValue : mergeValue && mergeValue[0], inputValue: inputValue, 
        // other
        isEmptyValue: isEmptyValue(mergeValue), prefixCls: prefixCls, isMultiple: isMultiple, renderText: renderText, onRemoveCheckedItem: onRemoveCheckedItem, onClear: function (e) {
            e.stopPropagation();
            if (!isMultiple) {
                handleChange([]);
            }
            else {
                var nodes = store.getCheckedNodes();
                var newValue = nodes.filter(function (x) { return x.disabled; }).map(function (x) { return x.pathValue; });
                store.setNodeCheckedByValue(newValue);
                handleChange(newValue, true);
            }
            props.onClear && props.onClear(!!popupVisible);
        }, onKeyDown: function (e) {
            if (disabled) {
                return;
            }
            e.stopPropagation();
            var keyCode = e.keyCode || e.which;
            if (keyCode === keycode_1.Enter.code && !popupVisible) {
                handleVisibleChange(true);
                e.preventDefault();
            }
            if (keyCode === keycode_1.Tab.code && popupVisible) {
                handleVisibleChange(false);
            }
        }, 
        // onFocus={this.onFocusInput}
        onChangeInputValue: function (v) {
            setInputValue(v);
            props.onSearch && props.onSearch(v);
            // tab键 focus 到输入框，此时下拉框未显示。如果输入值，展示下拉框
            if (!popupVisible) {
                handleVisibleChange(true);
            }
        } })))));
}
var CascaderComponent = (0, react_1.forwardRef)(Cascader);
CascaderComponent.displayName = 'Cascader';
exports.default = CascaderComponent;
