// @m-design/m-site-navbar/lib/utils/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLoginHref = exports.apiBasename = exports.isInternal = exports.isExternal = exports.hostname = exports.externalUrl = exports.internalUrl = void 0;
exports.internalUrl = "https://arco.bytedance.net";
exports.externalUrl = "https://arco.design";
exports.hostname = location.hostname !== "arco.design" &&
    location.hostname !== "arco.bytedance.net"
    ? "arco.design"
    : location.hostname;
exports.isExternal = exports.hostname === "arco.design";
exports.isInternal = exports.hostname === "arco.bytedance.net";
exports.apiBasename = "https://".concat(exports.hostname || "arco.design");
var getLoginHref = function () {
    // let host = "sso.meetsocial.cn";
    // const isProd = location.host === "m.meetsocial.cn";
    // return `https://${isProd ? "test-" : "dev-"}${host}?redirect=${encodeURIComponent(
    //   location.href
    // )}`;
    return "/login?redirect=".concat(encodeURIComponent(location.href));
};
exports.getLoginHref = getLoginHref;
