// @m-design/m-site-navbar/lib/components/select/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-nocheck
var react_1 = __importDefault(require("react"));
var web_react_1 = require("@arco-design/web-react");
var icon_1 = require("@arco-design/web-react/icon");
var style_module_less_1 = __importDefault(require("./style.module.less"));
function NavSelect(_a) {
    var value = _a.value, defaultValue = _a.defaultValue, options = _a.options, onChange = _a.onChange, triggerElement = _a.triggerElement, button = _a.button, buttonValue = _a.buttonValue, buttonIcon = _a.buttonIcon, onClickButton = _a.onClickButton;
    var selectProps = button
        ? {
            options: options.map(function (o) {
                var option = o;
                return {
                    label: react_1.default.createElement('span', null, react_1.default.createElement(icon_1.IconCheck, {
                        style: {
                            marginRight: 10,
                            fontSize: 12,
                            opacity: buttonValue === option.value ? 1 : 0,
                        },
                    }), option.label),
                    value: option.value,
                };
            }),
            triggerElement: react_1.default.createElement(web_react_1.Button, {
                className: style_module_less_1.default.btnIcon,
                type: 'text',
                icon: buttonIcon,
                size: 'mini',
                onClick: onClickButton,
            }),
        }
        : {};
    if (value) {
        selectProps.value = value;
    }
    else {
        selectProps.defaultValue = defaultValue;
    }
    return react_1.default.createElement(web_react_1.Select, Object.assign({
        trigger: 'hover',
        options: options,
        bordered: false,
        onChange: onChange,
        triggerProps: {
            autoAlignPopupWidth: false,
            autoAlignPopupMinWidth: true,
        },
        className: style_module_less_1.default.select,
        triggerElement: triggerElement,
    }, selectProps));
}
exports.default = NavSelect;
