// @m-design/m-site-navbar/es/locale.js
export default {
    'zh-CN': {
        search: '搜索',
        searchedResult: '搜索到 {0} 个结果',
        searchInArcoDesign: '在 Arco Design 中搜索...',
        noResult: '没有 {0} 的搜索结果',
        loading: '加载中',
        open: '打开',
        exit: '退出',
        component: '组件',
        material: '物料',
        theme: '主题',
        design: '设计',
        resource: '资源',
        about: '关于 M Design',
        modeDark: '暗黑模式',
        modeLight: '亮色模式',
        modeSystem: '跟随系统',
        joinChat: '加入 Oncall 群聊',
        needLogin: '未登录',
        loginNow: '立即登录',
        logout: '退出登录',
        myTheme: '我的主题',
        myMaterial: '我的物料',
        internalVersion: '跳转内网版',
        externalVersion: '跳转外网版',
        version: '版本',
        // components
        Icon: '图标 Icon',
        Button: '按钮 Button',
        Link: '链接 Link',
        Typography: '排版 Typography',
        Grid: '栅格 Grid',
        Divider: '分割线 Divider',
        Layout: '布局 Layout',
        Space: '间距 Space',
        Avatar: '头像 Avatar',
        Badge: '徽标 Badge',
        Calendar: '日历 Calendar',
        Card: '卡片 Card',
        Collapse: '折叠面板 Collapse',
        Comment: '评论 Comment',
        Descriptions: '描述列表 Descriptions',
        Empty: '空状态 Empty',
        List: '列表 List',
        Statistic: '数值显示 Statistic',
        Tabs: '标签页 Tabs',
        Table: '表格 Table',
        Tag: '标签 Tag',
        Timeline: '时间轴 Timeline',
        Tooltip: '文字气泡 Tooltip',
        Popover: '气泡卡片 Popover',
        Carousel: '图片轮播 Carousel',
        Tree: '树 Tree',
        Image: '图片 Image',
        DatePicker: '日期选择器 DatePicker',
        TimePicker: '时间选择器 TimePicker',
        Input: '输入框 Input',
        InputTag: '标签输入框 InputTag',
        InputNumber: '数字输入框 InputNumber',
        AutoComplete: '自动补全 AutoComplete',
        Checkbox: '复选框 Checkbox',
        Radio: '单选框 Radio',
        Rate: '评分 Rate',
        Switch: '开关 Switch',
        Select: '选择器 Select',
        TreeSelect: '树选择 TreeSelect',
        Cascader: '级联选择 Cascader',
        Slider: '滑动输入条 Slider',
        Form: '表单 Form',
        Upload: '上传 Upload',
        Transfer: '数据穿梭框 Transfer',
        Mentions: '提及 Mentions',
        Alert: '警告提示 Alert',
        Drawer: '抽屉 Drawer',
        Message: '全局提示 Message',
        Notification: '通知提醒框 Notification',
        Popconfirm: '气泡确认框 Popconfirm',
        Progress: '进度条 Progress',
        Result: '结果页 Result',
        Spin: '加载中 Spin',
        Modal: '对话框 Modal',
        Skeleton: '骨架屏 Skeleton',
        Breadcrumb: '面包屑 Breadcrumb',
        Dropdown: '下拉菜单 Dropdown',
        Menu: '菜单 Menu',
        PageHeader: '页头 PageHeader',
        Pagination: '分页 Pagination',
        Steps: '步骤条 Steps',
        ConfigProvider: '全局配置 ConfigProvider',
        Affix: '固钉 Affix',
        Anchor: '锚点 Anchor',
        BackTop: '返回顶部 BackTop',
        Trigger: '触发器 Trigger',
        ResizeBox: '伸缩框 ResizeBox',
    },
    'en-US': {
        search: 'Search',
        searchedResult: '{0} results found',
        searchInArcoDesign: 'Search in Arco Design...',
        noResult: 'No search results for {0}',
        loading: 'Loading',
        open: 'Open',
        exit: 'Exit',
        component: 'Component',
        material: 'Material',
        theme: 'Theme',
        design: 'Design',
        resource: 'Resource',
        modeDark: 'Dark Mode',
        modeLight: 'Light Mode',
        modeSystem: 'System theme',
        toLight: 'Light style',
        joinChat: 'OnCall group chat',
        needLogin: '401',
        loginNow: 'Login',
        logout: 'Logout',
        myTheme: 'My Themes',
        myMaterial: 'My Materials',
        internalVersion: 'Internal Docs',
        externalVersion: 'External Dos',
        version: 'Version',
        // components
        Icon: 'Icon',
        Button: 'Button',
        Link: 'Link',
        Typography: 'Typography',
        Grid: 'Grid',
        Divider: 'Divider',
        Layout: 'Layout',
        Space: 'Space',
        Avatar: 'Avatar',
        Badge: 'Badge',
        Calendar: 'Calendar',
        Card: 'Card',
        Collapse: 'Collapse',
        Comment: 'Comment',
        Descriptions: 'Descriptions',
        Empty: 'Empty',
        List: 'List',
        Statistic: 'Statistic',
        Tabs: 'Tabs',
        Table: 'Table',
        Tag: 'Tag',
        Timeline: 'Timeline',
        Tooltip: 'Tooltip',
        Popover: 'Popover',
        Carousel: 'Carousel',
        Tree: 'Tree',
        Image: 'Image',
        DatePicker: 'DatePicker',
        TimePicker: 'TimePicker',
        Input: 'Input',
        InputTag: 'InputTag',
        InputNumber: 'InputNumber',
        AutoComplete: 'AutoComplete',
        Checkbox: 'Checkbox',
        Radio: 'Radio',
        Rate: 'Rate',
        Switch: 'Switch',
        Select: 'Select',
        TreeSelect: 'TreeSelect',
        Cascader: 'Cascader',
        Slider: 'Slider',
        Form: 'Form',
        Upload: 'Upload',
        Transfer: 'Transfer',
        Mentions: 'Mentions',
        Alert: 'Alert',
        Drawer: 'Drawer',
        Message: 'Message',
        Notification: 'Notification',
        Popconfirm: 'Popconfirm',
        Progress: 'Progress',
        Result: 'Result',
        Spin: 'Spin',
        Modal: 'Modal',
        Skeleton: 'Skeleton',
        Breadcrumb: 'Breadcrumb',
        Dropdown: 'Dropdown',
        Menu: 'Menu',
        PageHeader: 'PageHeader',
        Pagination: 'Pagination',
        Steps: 'Steps',
        ConfigProvider: 'ConfigProvider',
        Affix: 'Affix',
        Anchor: 'Anchor',
        BackTop: 'BackTop',
        Trigger: 'Trigger',
        ResizeBox: 'ResizeBox',
    },
};
