// @arco-design/web-react/es/Typography/base.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
import React, { useState, useContext, useRef } from 'react';
import useIsomorphicLayoutEffect from '../_util/hooks/useIsomorphicLayoutEffect';
import { ConfigContext } from '../ConfigProvider';
import Operations from './operations';
import cs from '../_util/classNames';
import EditContent from './edit-content';
import { isObject } from '../_util/is';
import useResizeObserver from '../_util/hooks/useResizeObserver';
import { measure } from './utils';
import Tooltip from '../Tooltip';
import Popover from '../Popover';
import { raf, caf } from '../_util/raf';
import omit from '../_util/omit';
import useUpdateEffect from '../_util/hooks/useUpdate';
import mergedToString from '../_util/mergedToString';
function getClassNameAndComponentName(props, prefixCls) {
    var type = props.type, bold = props.bold, disabled = props.disabled, mark = props.mark, underline = props.underline, propDelete = props.delete, code = props.code;
    var component = [];
    var className = [];
    if (type) {
        className.push(prefixCls + "-" + type);
    }
    if (disabled) {
        className.push(prefixCls + "-disabled");
    }
    if (bold) {
        component.push('b');
    }
    if (underline) {
        component.push('u');
    }
    if (propDelete) {
        component.push('del');
    }
    if (code) {
        component.push('code');
    }
    if (mark) {
        component.push('mark');
    }
    return {
        component: component,
        className: className,
    };
}
function wrap(content, component, props) {
    var currentContent = content;
    component.forEach(function (c) {
        var _props = isObject(props.mark) && props.mark.color
            ? { style: { backgroundColor: props.mark.color } }
            : {};
        currentContent = React.createElement(c, _props, currentContent);
    });
    return currentContent;
}
function Base(props) {
    var componentType = props.componentType, style = props.style, className = props.className, children = props.children, editable = props.editable, ellipsis = props.ellipsis, heading = props.heading, blockquote = props.blockquote, rest = __rest(props, ["componentType", "style", "className", "children", "editable", "ellipsis", "heading", "blockquote"]);
    var getPrefixCls = useContext(ConfigContext).getPrefixCls;
    var prefixCls = getPrefixCls('typography');
    var rafId = useRef();
    var _a = getClassNameAndComponentName(props, prefixCls), component = _a.component, componentClassName = _a.className;
    var _b = __read(useState(false), 2), editing = _b[0], setEditing = _b[1];
    var _c = __read(useState(false), 2), isEllipsis = _c[0], setEllipsis = _c[1];
    var _d = __read(useState(false), 2), expanding = _d[0], setExpanding = _d[1];
    var _e = __read(useState(''), 2), ellipsisText = _e[0], setEllipsisText = _e[1];
    var _f = __read(useState(false), 2), measuring = _f[0], setMeasuring = _f[1];
    var componentRef = useRef(null);
    var editableConfig = isObject(editable) ? editable : {};
    var mergedEditing = 'editing' in editableConfig ? editableConfig.editing : editing;
    var ellipsisConfig = ellipsis
        ? __assign({ rows: 1, ellipsisStr: '...' }, (isObject(ellipsis) ? ellipsis : {})) : {};
    function renderOperations(forceShowExpand) {
        return (React.createElement(React.Fragment, null,
            React.createElement(Operations, __assign({}, props, { setEditing: setEditing, onClickExpand: onClickExpand, expanding: expanding, isEllipsis: isEllipsis, forceShowExpand: forceShowExpand }))));
    }
    function onClickExpand() {
        setExpanding(!expanding);
        ellipsisConfig.onExpand && ellipsisConfig.onExpand(!expanding);
    }
    var resizeOnNextFrame = function () {
        caf(rafId.current);
        rafId.current = raf(function () {
            calcEllipsis();
        });
    };
    var _g = useResizeObserver(resizeOnNextFrame), cor = _g.cor, dor = _g.dor, currentOr = _g.currentOr;
    useIsomorphicLayoutEffect(function () {
        if (!currentOr) {
            resizeOnNextFrame();
        }
    }, [currentOr]);
    useUpdateEffect(function () {
        ellipsisConfig.onEllipsis && ellipsisConfig.onEllipsis(isEllipsis);
    }, [isEllipsis]);
    useIsomorphicLayoutEffect(function () {
        if (componentRef.current) {
            cor(componentRef.current);
        }
        return function () {
            dor();
            caf(rafId.current);
        };
    }, [
        children,
        expanding,
        isEllipsis,
        editing,
        ellipsisConfig.suffix,
        ellipsisConfig.ellipsisStr,
        ellipsisConfig.expandable,
        ellipsisConfig.expandNodes,
        ellipsisConfig.rows,
    ]);
    function calcEllipsis() {
        var currentEllipsis = isEllipsis;
        if (editing) {
            return;
        }
        if (ellipsisConfig.rows) {
            // In ellipsis mode, if the user manually expands, there is no need to calculate ellipsis and ellipsisText;
            if (expanding) {
                return;
            }
            setMeasuring(true);
            var _a = measure(componentRef.current, ellipsisConfig, renderOperations(!!ellipsisConfig.expandable), children), ellipsis_1 = _a.ellipsis, text = _a.text;
            setMeasuring(false);
            if (ellipsis_1 && text) {
                setEllipsisText(text);
            }
            if (ellipsis_1 !== currentEllipsis) {
                setEllipsis(ellipsis_1);
            }
        }
        else {
            var isEllipsis_1 = !!ellipsisConfig.rows;
            if (isEllipsis_1 !== currentEllipsis) {
                setEllipsis(isEllipsis_1);
            }
        }
    }
    function renderContent() {
        var fullText = mergedToString(React.Children.toArray(children));
        var ellipsisStr = ellipsisConfig.ellipsisStr !== undefined ? ellipsisConfig.ellipsisStr : '...';
        var suffix = ellipsisConfig.suffix !== undefined && ellipsisConfig.suffix;
        var showTooltip = ellipsisConfig.showTooltip;
        var tooltipType = isObject(ellipsisConfig.showTooltip)
            ? ellipsisConfig.showTooltip.type === 'popover'
                ? 'popover'
                : 'tooltip'
            : 'tooltip';
        var tooltipProps = isObject(ellipsisConfig.showTooltip)
            ? ellipsisConfig.showTooltip.props || {}
            : {};
        var TooltipComponent = (tooltipType === 'popover' ? Popover : Tooltip);
        var titleProps = isEllipsis && !showTooltip && !expanding ? { title: fullText } : {};
        var baseProps = __assign({ ref: componentRef, style: style }, titleProps);
        function renderInnerContent() {
            var text = isEllipsis && !expanding ? ellipsisText : children;
            var innerText = component.length ? wrap(text, component, props) : text;
            return (React.createElement(React.Fragment, null,
                isEllipsis && showTooltip && !expanding ? (React.createElement(TooltipComponent, __assign({ content: fullText }, tooltipProps), innerText)) : (innerText),
                measuring || (isEllipsis && !expanding) ? ellipsisStr : null,
                suffix,
                renderOperations(measuring ? !!ellipsisConfig.expandable : undefined)));
        }
        var TextComponent;
        if (componentType === 'Paragraph') {
            TextComponent = blockquote ? 'blockquote' : 'div';
        }
        else if (componentType === 'Title') {
            TextComponent = "h" + heading;
        }
        else if (componentType === 'Text') {
            TextComponent = ellipsis ? 'div' : 'span';
        }
        return (React.createElement(TextComponent, __assign({ className: cs(prefixCls, componentClassName, className) }, baseProps, omit(rest, [
            'spacing',
            'type',
            'close',
            'bold',
            'disabled',
            'mark',
            'underline',
            'delete',
            'code',
            'copyable',
            'isEllipsis',
            'expanding',
            'onClickExpand',
            'setEditing',
            'forceShowExpand',
        ])), renderInnerContent()));
    }
    return mergedEditing ? (React.createElement(EditContent, __assign({}, props, { prefixCls: prefixCls, setEditing: setEditing, editableConfig: editableConfig }))) : (renderContent());
}
export default Base;
