// @m-design/m-site-navbar/lib/components/dropdown/about.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-nocheck
var react_1 = __importStar(require("react"));
var g2_1 = require("@antv/g2");
// import axios from 'axios';
// import dayjs from 'dayjs';
var icon_1 = __importDefault(require("../../icon"));
var about_module_less_1 = __importDefault(require("./style/about.module.less"));
// import { apiBasename } from '../../utils/config';
var locales = {
    "zh-CN": {
        "download-day": "今日下载量",
        "download-all": "总下载量",
        dashboard: "Arco Dashboard",
    },
    "en-US": {
        "download-day": "Today Download",
        "download-all": "All Download",
        dashboard: "Arco Dashboard",
    },
};
function renderChart(container, data) {
    var chart = new g2_1.Chart({
        container: container,
        autoFit: true,
    });
    chart.data(data);
    chart.tooltip({
        showCrosshairs: true,
        containerTpl: '<div class="g2-tooltip" style="white-space: nowrap;padding-bottom: 20px">' +
            '<div class="g2-tooltip-title"></div>' +
            '<div class="g2-tooltip-list"></div></div>',
        itemTpl: '<div style="margin-bottom: 10px">{name}: {value}</div>',
        position: "top",
    });
    chart.axis(false); // 不展示坐标轴
    chart.line().position("endDate*count").shape("smooth");
    chart.area().position("endDate*count").shape("smooth");
    chart.render();
}
function About(_a) {
    var _b = _a.lang, lang = _b === void 0 ? "zh-CN" : _b;
    var _c = (0, react_1.useReducer)(function (v) {
        return v + 1;
    }, 0);
    // const forceUpdate = _c[1];
    var npmData = window.arcoNpmData || {};
    var dayBuckets = npmData.dayBuckets ? npmData.dayBuckets : [];
    var monthBuckets = npmData.monthBuckets ? npmData.monthBuckets : [];
    (0, react_1.useEffect)(function () {
        if (!npmData || !Object.keys(npmData).length) {
            getData();
        }
        else {
            renderChart("chart1", dayBuckets);
            renderChart("chart2", monthBuckets);
        }
    }, []);
    function getData() {
        //
    }
    var t = locales[lang];
    return (react_1.default.createElement("div", { className: about_module_less_1.default.about },
        react_1.default.createElement("div", { className: about_module_less_1.default.container },
            react_1.default.createElement("div", { className: about_module_less_1.default.download },
                react_1.default.createElement("div", { className: about_module_less_1.default.chart1, id: "chart1" }),
                react_1.default.createElement("div", { className: about_module_less_1.default.text },
                    react_1.default.createElement("div", { className: about_module_less_1.default.count }, npmData.today ? npmData.today.toLocaleString() : "..."),
                    react_1.default.createElement("div", { className: about_module_less_1.default.desc }, t["download-day"]))),
            react_1.default.createElement("div", { className: about_module_less_1.default.download },
                react_1.default.createElement("div", { className: about_module_less_1.default.chart2, id: "chart2" }),
                react_1.default.createElement("div", { className: about_module_less_1.default.text },
                    react_1.default.createElement("div", { className: about_module_less_1.default.count }, npmData.all ? npmData.all.toLocaleString() : "..."),
                    react_1.default.createElement("div", { className: about_module_less_1.default.desc }, t["download-all"]))),
            react_1.default.createElement("a", { className: about_module_less_1.default.link, href: "/dashboard", target: "_blank", rel: "noreferrer" },
                react_1.default.createElement("div", { className: about_module_less_1.default.icon }, icon_1.default.IconDashboard),
                t.dashboard))));
}
exports.default = About;
