// @m-design/m-site-navbar/lib/components/dropdown/development.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-nocheck
var react_1 = __importStar(require("react"));
var web_react_1 = require("@arco-design/web-react");
var gsap_1 = __importDefault(require("gsap"));
var nav_ic_react_svg_1 = __importDefault(require("../../assets/nav_ic_react.svg"));
// import IconVue from '../../assets/nav_ic_vue.svg';
var development_module_less_1 = __importDefault(require("./style/development.module.less"));
var icon_1 = __importDefault(require("../../icon"));
var locales = {
    "zh-CN": {
        docs: "开发文档",
        resource: "开发资源",
        toolMigrate: "迁移工具 - 迁移到 ArcoDesign 2.0",
        toolArcoCli: "M CLI - 官方脚手架",
        toolWebpack: "M Design 官方 webpack 插件",
        github: "代码仓库",
    },
    "en-US": {
        docs: "Development Docs",
        resource: "Development Resources",
        toolMigrate: "Code migrate to ArcoDesign 2.0",
        toolArcoCli: "Arco CLI - official scaffold tool",
        toolWebpack: "Official Webpack Plugin",
        github: "GitHub repository",
    },
};
var library = [
    {
        title: "React",
        key: "react",
        npm: "@arco-design/web-react",
        link: "/react/docs/start",
        svg: react_1.default.createElement(nav_ic_react_svg_1.default, null),
    },
    // {
    //     title: 'Vue',
    //     key: 'vue',
    //     npm: '@arco-design/web-vue',
    //     link: '/vue',
    //     svg: React.createElement(IconVue, null),
    // },
];
var devResource = [
    {
        title: "github",
        link: "http://192.168.1.6:8088/m-design/m-design",
        svg: icon_1.default.IconGitHub,
    },
    {
        title: "toolArcoCli",
        link: "http://192.168.1.6:8088/m-design/m-design",
        svg: icon_1.default.IconCli,
    },
    {
        title: "toolWebpack",
        link: "http://192.168.1.6:8088/m-design/m-design",
        svg: icon_1.default.IconWebpack,
    },
];
function Component(_a) {
    var _b = _a.lang, lang = _b === void 0 ? "zh-CN" : _b;
    var _c = (0, react_1.useReducer)(function (v) {
        return v + 1;
    }, 0), forceUpdate = _c[1];
    var version = window.navVersion || [];
    (0, react_1.useEffect)(function () {
        if (!version || !version.length) {
            // eslint-disable-next-line @typescript-eslint/no-use-before-define
            getVersions();
        }
    }, []);
    function getVersions() {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_d) {
                return [2 /*return*/];
            });
        });
    }
    (0, react_1.useEffect)(function () {
        var timeline = gsap_1.default.timeline();
        timeline
            .from(".".concat(development_module_less_1.default.left, ", .").concat(development_module_less_1.default.right), {
            opacity: 0,
            duration: 0.3,
            x: -20,
            ease: "cubic-bezier(0.25, 1, 0.5, 1)",
            stagger: {
                amount: 0.1,
            },
            delay: 0.2,
        })
            .from(".".concat(development_module_less_1.default.componentItem), {
            opacity: 0,
            duration: 0.35,
            ease: "cubic-bezier(0.25, 1, 0.5, 1)",
            x: -16,
            stagger: {
                amount: 0.2,
                grid: [5, 4],
                from: "start",
                axis: "x",
            },
        }, "<0.2");
    }, []);
    var t = locales[lang];
    return (react_1.default.createElement("div", { className: development_module_less_1.default.development },
        react_1.default.createElement(web_react_1.Grid.Row, { className: development_module_less_1.default.container, gutter: 20 },
            react_1.default.createElement(web_react_1.Grid.Col, { span: 12, className: development_module_less_1.default.left },
                react_1.default.createElement("div", { className: development_module_less_1.default.title }, t.docs),
                react_1.default.createElement("div", { className: development_module_less_1.default.list }, library.map(function (l, index) { return (react_1.default.createElement("a", { className: development_module_less_1.default.item, href: l.link, key: index, target: "_blank", rel: "noreferrer" },
                    l.svg,
                    l.title)); }))),
            react_1.default.createElement(web_react_1.Grid.Col, { span: 12, className: development_module_less_1.default.right },
                react_1.default.createElement("div", { className: development_module_less_1.default.title }, t.resource),
                react_1.default.createElement("div", { className: development_module_less_1.default.list }, devResource.map(function (d, index) { return (react_1.default.createElement("a", { className: development_module_less_1.default.item, href: d.link, key: index, target: "_blank", rel: "noreferrer" },
                    react_1.default.createElement("div", { className: development_module_less_1.default.icon }, d.svg),
                    t[d.title])); }))))));
}
exports.default = Component;
