// @m-design/m-site-navbar/lib/components/dropdown/ecosystem.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-nocheck
var react_1 = __importStar(require("react"));
var gsap_1 = __importDefault(require("gsap"));
var web_react_1 = require("@arco-design/web-react");
var ecosystem_module_less_1 = __importDefault(require("./style/ecosystem.module.less"));
var icon_1 = __importDefault(require("../../icon"));
var config_1 = require("../../utils/config");
var locales = {
    "zh-CN": {
        pro: "M Pro 中后台最佳实践",
        proDescription: "快速构建中后台前端解决方案",
        designLab: "DesignLab 风格配置平台",
        designLabDescription: "精确到组件级的主题视觉配置",
        material: "Material Market 物料平台",
        materialDescription: "丰富可共享的业务定制物料，让效率突破猛进",
        iconBox: "IconBox 图标平台",
        iconBoxDescription: "一站式图标管理平台",
        palette: "Palette 色彩配置工具",
        paletteDescription: "色彩配置从未如此简单",
        chartSpace: "ChartSpace 图表库",
        chartSpaceDescription: "简单、易用、跨平台、高性能的图表库与研发框架",
    },
    "en-US": {
        pro: "ArcoPro - Best Practices",
        proDescription: "A solution to quickly building applications ",
        designLab: "Design System Lab",
        designLabDescription: "Flexible theme customization, even available at component level",
        material: "Component Market",
        materialDescription: "Provide massive high-quality customized materials that can greatly boost efficiency",
        iconBox: "IconBox",
        iconBoxDescription: "One-stop platform to manage your icons",
        palette: "Palette",
        paletteDescription: "Color setting can never be easier",
        chartSpace: "ChartSpace",
        chartSpaceDescription: "Simple, easy-to-use, cross-platform, high-performance chart library and R&D framework",
    },
};
var ecosystems = [
    {
        title: 'pro',
        description: 'proDescription',
        link: '/pro',
        svg: icon_1.default.IconLogoArcoProW,
    },
    {
        title: "designLab",
        description: "designLabDescription",
        link: "/themes",
        svg: icon_1.default.IconLogoDesignLabW,
    },
    {
        title: "material",
        description: "materialDescription",
        link: "/material",
        svg: icon_1.default.IconLogoArcoMaterialW,
    },
    // {
    //     title: 'iconBox',
    //     description: 'iconBoxDescription',
    //     link: '/iconbox',
    //     svg: icons.IconLogoIconBoxW,
    // },
];
if (config_1.isInternal) {
    ecosystems.push({
        title: "chartSpace",
        description: "chartSpaceDescription",
        link: "https://chartspace.web.bytedance.net/home",
        svg: icon_1.default.IconChartSpaceW,
    });
}
function Resource(_a) {
    var _b = _a.lang, lang = _b === void 0 ? "zh-CN" : _b;
    (0, react_1.useEffect)(function () {
        var timeline = gsap_1.default.timeline();
        timeline.from(".".concat(ecosystem_module_less_1.default.item), {
            opacity: 0,
            duration: 0.35,
            ease: "cubic-bezier(0.25, 1, 0.5, 1)",
            x: -16,
            stagger: {
                amount: 0.2,
                from: "start",
                grid: [3, 3],
                axis: "x",
            },
        }, "<0.2");
    }, []);
    var t = locales[lang];
    function toBeOpen(e, url) {
        if (!url) {
            e.preventDefault();
            web_react_1.Message.info(t.toBeOpen);
        }
    }
    return (react_1.default.createElement("div", { className: ecosystem_module_less_1.default.ecosystem },
        react_1.default.createElement("div", { className: ecosystem_module_less_1.default.container },
            react_1.default.createElement("div", { className: ecosystem_module_less_1.default.content }, ecosystems.map(function (d, index) { return (react_1.default.createElement("a", { className: ecosystem_module_less_1.default.item, href: d.link, key: index, target: "_blank", rel: "noreferrer", onClick: function () { return toBeOpen(e, d.link); } },
                react_1.default.createElement("div", { className: ecosystem_module_less_1.default.icon }, d.svg),
                react_1.default.createElement("div", { className: ecosystem_module_less_1.default.wrapper },
                    react_1.default.createElement("div", { className: ecosystem_module_less_1.default.title }, t[d.title]),
                    react_1.default.createElement("div", { className: ecosystem_module_less_1.default.description }, t[d.description])))); })))));
}
exports.default = Resource;
