// @m-design/m-site-navbar/lib/components/dropdown/spec.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// @ts-nocheck
var react_1 = __importStar(require("react"));
var web_react_1 = require("@arco-design/web-react");
var gsap_1 = __importDefault(require("gsap"));
var spec_module_less_1 = __importDefault(require("./style/spec.module.less"));
var icon_1 = __importDefault(require("../../icon"));
var locales = {
    "zh-CN": {
        introduce: "M Design 简介",
        componentUsage: "组件用法",
        designValues: "设计价值观",
        designPhilosophy: "设计原则",
        styleGuideline: "样式指南",
        designGuideline: "设计指南",
        designResource: "设计资源",
        componentFigma: "Figma 组件库",
        proFigma: "ArcoPro 中后台最佳实践",
        iconFigma: "官方图标库",
        sketchResource: "Sketch 组件库",
        axureResource: "Axure 组件库",
    },
    "en-US": {
        introduce: "ArcoDesign Introduce",
        componentUsage: "Component Usage",
        designValues: "Design Values",
        designPhilosophy: "Design Philosophy",
        styleGuideline: "Style Guideline",
        designGuideline: "Design Guideline",
        designResource: "Design Resources",
        componentFigma: "Figma Component Library ",
        proFigma: "ArcoPro Best Practice Resource",
        iconFigma: "Official Icon Library",
        sketchResource: "Sketch Resource",
        axureResource: "Axure Resource",
    },
};
var designGuidelines = [
    {
        title: "introduce",
        link: "/docs/spec/introduce",
        svg: icon_1.default.IconArcoIntroduce,
    },
    {
        title: "componentUsage",
        link: "/docs/spec/link",
        svg: icon_1.default.IconComponentUsage,
    },
    {
        title: "designValues",
        link: "/docs/spec/values-of-arcodesign",
        svg: icon_1.default.IconDesignValues,
    },
    // {
    //   title: 'designPhilosophy',
    //   link: '/docs/spec/philosophy',
    //   svg: icons.IconDesignPhilosophy,
    // },
    // {
    //   title: 'styleGuideline',
    //   link: '/docs/spec/style-guideline',
    //   svg: icons.IconStyleGuideline,
    // },
];
var designResource = [
    // {
    //   title: 'componentFigma',
    //   link:
    //     'https://www.figma.com/file/M66cTiLXHa4SVyZIlfY5Pb/arco-Design-System?node-id=7945%3A44563',
    //   svg: icons.IconFigmaWhite,
    // },
    // {
    //   title: 'proFigma',
    //   link: 'https://www.figma.com/file/dp8okiO9c6tKdBmqv1m1R2/Arco-Design-Pro?node-id=515%3A5594',
    //   svg: icons.IconArcoProWhite,
    // },
    {
        title: "sketchResource",
        link: "https://s3.meetsocial.cn/mdesign/assets/MDesign.sketch",
        svg: icon_1.default.IconSketch,
    },
    // {
    //   title: 'iconFigma',
    //   link: 'https://www.figma.com/file/1ohmb16op4ogbI09ojLR5W/Arco-Design-IconsFigma',
    //   svg: icons.IconIconWhite,
    // },
    {
        title: "axureResource",
        link: "https://s3.meetsocial.cn/mdesign/assets/MDesign_Axure.rplib",
        svg: icon_1.default.IconAxure,
    },
];
function Spec(_a) {
    var _b = _a.lang, lang = _b === void 0 ? "zh-CN" : _b;
    (0, react_1.useEffect)(function () {
        var timeline = gsap_1.default.timeline();
        timeline
            .from(".".concat(spec_module_less_1.default.left, ", .").concat(spec_module_less_1.default.right), {
            opacity: 0,
            duration: 0.3,
            x: -20,
            ease: "cubic-bezier(0.25, 1, 0.5, 1)",
            stagger: {
                amount: 0.1,
            },
            delay: 0.2,
        })
            .from(".".concat(spec_module_less_1.default.componentItem), {
            opacity: 0,
            duration: 0.35,
            ease: "cubic-bezier(0.25, 1, 0.5, 1)",
            x: -16,
            stagger: {
                amount: 0.2,
                grid: [5, 4],
                from: "start",
                axis: "x",
            },
        }, "<0.2");
    }, []);
    var t = locales[lang];
    return (react_1.default.createElement("div", { className: spec_module_less_1.default.spec },
        react_1.default.createElement(web_react_1.Grid.Row, { className: spec_module_less_1.default.container, gutter: 20 },
            react_1.default.createElement(web_react_1.Grid.Col, { span: 12, className: spec_module_less_1.default.left },
                react_1.default.createElement("div", { className: spec_module_less_1.default.title }, t.designGuideline),
                react_1.default.createElement("div", { className: spec_module_less_1.default.list }, designGuidelines.map(function (d, index) { return (react_1.default.createElement("a", { className: spec_module_less_1.default.item, href: d.link, key: index, target: "_blank", rel: "noreferrer" },
                    react_1.default.createElement("div", { className: spec_module_less_1.default.icon }, d.svg),
                    t[d.title])); }))),
            react_1.default.createElement(web_react_1.Grid.Col, { span: 12, className: spec_module_less_1.default.right },
                react_1.default.createElement("div", { className: spec_module_less_1.default.title }, t.designResource),
                react_1.default.createElement("div", { className: spec_module_less_1.default.list }, designResource.map(function (d, index) { return (react_1.default.createElement("a", { className: spec_module_less_1.default.item, href: d.link, key: index, target: "_blank", rel: "noreferrer" },
                    react_1.default.createElement("div", { className: spec_module_less_1.default.icon }, d.svg),
                    t[d.title])); }))))));
}
exports.default = Spec;
