// @m-design/m-site-navbar/es/components/dropdown/ecosystem.js
// @ts-nocheck
import React, { useEffect } from "react";
import gsap from "gsap";
import { Message } from "@arco-design/web-react";
import styles from "./style/ecosystem.module.less";
import icons from "../../icon";
import { isInternal } from "../../utils/config";
var locales = {
    "zh-CN": {
        pro: "M Pro 中后台最佳实践",
        proDescription: "快速构建中后台前端解决方案",
        designLab: "DesignLab 风格配置平台",
        designLabDescription: "精确到组件级的主题视觉配置",
        material: "Material Market 物料平台",
        materialDescription: "丰富可共享的业务定制物料，让效率突破猛进",
        iconBox: "IconBox 图标平台",
        iconBoxDescription: "一站式图标管理平台",
        palette: "Palette 色彩配置工具",
        paletteDescription: "色彩配置从未如此简单",
        chartSpace: "ChartSpace 图表库",
        chartSpaceDescription: "简单、易用、跨平台、高性能的图表库与研发框架",
    },
    "en-US": {
        pro: "ArcoPro - Best Practices",
        proDescription: "A solution to quickly building applications ",
        designLab: "Design System Lab",
        designLabDescription: "Flexible theme customization, even available at component level",
        material: "Component Market",
        materialDescription: "Provide massive high-quality customized materials that can greatly boost efficiency",
        iconBox: "IconBox",
        iconBoxDescription: "One-stop platform to manage your icons",
        palette: "Palette",
        paletteDescription: "Color setting can never be easier",
        chartSpace: "ChartSpace",
        chartSpaceDescription: "Simple, easy-to-use, cross-platform, high-performance chart library and R&D framework",
    },
};
var ecosystems = [
    {
        title: 'pro',
        description: 'proDescription',
        link: '/pro',
        svg: icons.IconLogoArcoProW,
    },
    {
        title: "designLab",
        description: "designLabDescription",
        link: "/themes",
        svg: icons.IconLogoDesignLabW,
    },
    {
        title: "material",
        description: "materialDescription",
        link: "/material",
        svg: icons.IconLogoArcoMaterialW,
    },
    // {
    //     title: 'iconBox',
    //     description: 'iconBoxDescription',
    //     link: '/iconbox',
    //     svg: icons.IconLogoIconBoxW,
    // },
];
if (isInternal) {
    ecosystems.push({
        title: "chartSpace",
        description: "chartSpaceDescription",
        link: "https://chartspace.web.bytedance.net/home",
        svg: icons.IconChartSpaceW,
    });
}
function Resource(_a) {
    var _b = _a.lang, lang = _b === void 0 ? "zh-CN" : _b;
    useEffect(function () {
        var timeline = gsap.timeline();
        timeline.from(".".concat(styles.item), {
            opacity: 0,
            duration: 0.35,
            ease: "cubic-bezier(0.25, 1, 0.5, 1)",
            x: -16,
            stagger: {
                amount: 0.2,
                from: "start",
                grid: [3, 3],
                axis: "x",
            },
        }, "<0.2");
    }, []);
    var t = locales[lang];
    function toBeOpen(e, url) {
        if (!url) {
            e.preventDefault();
            Message.info(t.toBeOpen);
        }
    }
    return (React.createElement("div", { className: styles.ecosystem },
        React.createElement("div", { className: styles.container },
            React.createElement("div", { className: styles.content }, ecosystems.map(function (d, index) { return (React.createElement("a", { className: styles.item, href: d.link, key: index, target: "_blank", rel: "noreferrer", onClick: function () { return toBeOpen(e, d.link); } },
                React.createElement("div", { className: styles.icon }, d.svg),
                React.createElement("div", { className: styles.wrapper },
                    React.createElement("div", { className: styles.title }, t[d.title]),
                    React.createElement("div", { className: styles.description }, t[d.description])))); })))));
}
export default Resource;
