// @m-design/m-site-navbar/es/components/dropdown/spec.js
// @ts-nocheck
import React, { useEffect } from "react";
import { Grid } from "@arco-design/web-react";
import gsap from "gsap";
import styles from "./style/spec.module.less";
import icons from "../../icon";
var locales = {
    "zh-CN": {
        introduce: "M Design 简介",
        componentUsage: "组件用法",
        designValues: "设计价值观",
        designPhilosophy: "设计原则",
        styleGuideline: "样式指南",
        designGuideline: "设计指南",
        designResource: "设计资源",
        componentFigma: "Figma 组件库",
        proFigma: "ArcoPro 中后台最佳实践",
        iconFigma: "官方图标库",
        sketchResource: "Sketch 组件库",
        axureResource: "Axure 组件库",
    },
    "en-US": {
        introduce: "ArcoDesign Introduce",
        componentUsage: "Component Usage",
        designValues: "Design Values",
        designPhilosophy: "Design Philosophy",
        styleGuideline: "Style Guideline",
        designGuideline: "Design Guideline",
        designResource: "Design Resources",
        componentFigma: "Figma Component Library ",
        proFigma: "ArcoPro Best Practice Resource",
        iconFigma: "Official Icon Library",
        sketchResource: "Sketch Resource",
        axureResource: "Axure Resource",
    },
};
var designGuidelines = [
    {
        title: "introduce",
        link: "/docs/spec/introduce",
        svg: icons.IconArcoIntroduce,
    },
    {
        title: "componentUsage",
        link: "/docs/spec/link",
        svg: icons.IconComponentUsage,
    },
    {
        title: "designValues",
        link: "/docs/spec/values-of-arcodesign",
        svg: icons.IconDesignValues,
    },
    // {
    //   title: 'designPhilosophy',
    //   link: '/docs/spec/philosophy',
    //   svg: icons.IconDesignPhilosophy,
    // },
    // {
    //   title: 'styleGuideline',
    //   link: '/docs/spec/style-guideline',
    //   svg: icons.IconStyleGuideline,
    // },
];
var designResource = [
    // {
    //   title: 'componentFigma',
    //   link:
    //     'https://www.figma.com/file/M66cTiLXHa4SVyZIlfY5Pb/arco-Design-System?node-id=7945%3A44563',
    //   svg: icons.IconFigmaWhite,
    // },
    // {
    //   title: 'proFigma',
    //   link: 'https://www.figma.com/file/dp8okiO9c6tKdBmqv1m1R2/Arco-Design-Pro?node-id=515%3A5594',
    //   svg: icons.IconArcoProWhite,
    // },
    {
        title: "sketchResource",
        link: "https://s3.meetsocial.cn/mdesign/assets/MDesign.sketch",
        svg: icons.IconSketch,
    },
    // {
    //   title: 'iconFigma',
    //   link: 'https://www.figma.com/file/1ohmb16op4ogbI09ojLR5W/Arco-Design-IconsFigma',
    //   svg: icons.IconIconWhite,
    // },
    {
        title: "axureResource",
        link: "https://s3.meetsocial.cn/mdesign/assets/MDesign_Axure.rplib",
        svg: icons.IconAxure,
    },
];
function Spec(_a) {
    var _b = _a.lang, lang = _b === void 0 ? "zh-CN" : _b;
    useEffect(function () {
        var timeline = gsap.timeline();
        timeline
            .from(".".concat(styles.left, ", .").concat(styles.right), {
            opacity: 0,
            duration: 0.3,
            x: -20,
            ease: "cubic-bezier(0.25, 1, 0.5, 1)",
            stagger: {
                amount: 0.1,
            },
            delay: 0.2,
        })
            .from(".".concat(styles.componentItem), {
            opacity: 0,
            duration: 0.35,
            ease: "cubic-bezier(0.25, 1, 0.5, 1)",
            x: -16,
            stagger: {
                amount: 0.2,
                grid: [5, 4],
                from: "start",
                axis: "x",
            },
        }, "<0.2");
    }, []);
    var t = locales[lang];
    return (React.createElement("div", { className: styles.spec },
        React.createElement(Grid.Row, { className: styles.container, gutter: 20 },
            React.createElement(Grid.Col, { span: 12, className: styles.left },
                React.createElement("div", { className: styles.title }, t.designGuideline),
                React.createElement("div", { className: styles.list }, designGuidelines.map(function (d, index) { return (React.createElement("a", { className: styles.item, href: d.link, key: index, target: "_blank", rel: "noreferrer" },
                    React.createElement("div", { className: styles.icon }, d.svg),
                    t[d.title])); }))),
            React.createElement(Grid.Col, { span: 12, className: styles.right },
                React.createElement("div", { className: styles.title }, t.designResource),
                React.createElement("div", { className: styles.list }, designResource.map(function (d, index) { return (React.createElement("a", { className: styles.item, href: d.link, key: index, target: "_blank", rel: "noreferrer" },
                    React.createElement("div", { className: styles.icon }, d.svg),
                    t[d.title])); }))))));
}
export default Spec;
